```{r eval=FALSE}


library(Seurat)
library(tidyverse)
library(cowplot)
library(Matrix)
library(viridis)
library(harmony)
library(ggpubr)
library(patchwork)
library(RColorBrewer)
library(ggrepel)
library(ggrastr)
theme_set(theme_cowplot())

setwd("/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/")

fig_dir <- "figures/"
data_dir <- "data/"



umap_theme <- theme(
  axis.line=element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  axis.ticks=element_blank(),
  axis.title.x=element_blank(),
  axis.title.y=element_blank(),
  panel.background=element_blank(),
  panel.border=element_blank(),
  panel.grid.major=element_blank(),
  panel.grid.minor=element_blank(),
  plot.background=element_blank()
)


# re-load integrated
seurat_obj <- readRDS(file="/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_AD_integrated.rds" )

seurat_adds <- readRDS(file="/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_integrated.rds" )

tic(); dim(subset(seurat_adds, cell_type == 'ASC')); toc()

# re-load ADDS seurat object:
seurat_AD <- readRDS(file="/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/AD_integrated.rds" )

var_df <- read.delim(file = "/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_subclusters_var.tsv", sep='\t')

VariableFeatures(seurat_adds) <- var_df$X
VariableFeatures(seurat_AD) <- var_df$X
VariableFeatures(seurat_obj) <- var_df$X

# load gene name table from split-pipe:
gene_table <- read.csv("/dfs3b/swaruplab/smorabit/data/ADDS_2021/analysis/Batch1/Sublibrary1_S1/all-well/DGE_filtered/genes.csv")

library(EnsDb.Hsapiens.v86)
library(GenomicRanges)
library(ensembldb)

# load DSAD vs Control DEGs:
degs2 <- read.csv('/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/DEGs/data/DSAD_vs_Control_celltype_DEGs_full.csv')



subset(degs, gene == 'AQP4') %>% head
subset(degs, gene == 'S100B') %>% head


```

function to get a certain chromosome and plot the DEGs

```{r eval=FALSE}

chr_names <- as.character(1:21)

PlotDEGsChromosome <- function(
  degs,
  gene_table, # matches the gene names in teh seurat object with ensembl ids
  EnsDb,
  chr_names,
  gene_biotypes = c("protein_coding")
){

  logfc_limit <- 0.5

  degs <- read.csv('/dfs3b/swaruplab/smorabit/analysis/ADDS_2021/splitseq/DEGs/data/DSAD_vs_Control_celltype_DEGs_full.csv')

  degs$avg_log2FC <- ifelse(
    abs(degs$avg_log2FC) > logfc_limit,
    sign(degs$avg_log2FC) * logfc_limit,
    degs$avg_log2FC
  )

  # genes that are differentially expressed in at least one cell type:
  logfc_thresh <- 0.25
  genes.keep <- unique(degs[abs(degs$avg_log2FC) >= logfc_thresh, 'gene'])
  degs <- subset(degs, gene %in% genes.keep)

  plot_list <- list()
  chr_n_genes <- c()

  for(cur_chr in chr_names){
    print(cur_chr)

    # subset the ensembl db based on
    gene.coords <- ensembldb::genes(EnsDb, filter = ~ gene_biotype == "protein_coding") %>% subset(seqnames == cur_chr)

    # subset by genes in our snRNA data
    gene.coords <- gene.coords[gene.coords$gene_id %in% gene_table$gene_id]
    ix <- match(gene.coords$gene_id, gene_table$gene_id)
    gene.coords$gene_name <- gene_table$gene_name[ix]

    # remove entries with duplicate names
    dupes <- names(which(table(gene.coords$gene_name) > 1))
    dupes <- unlist(lapply(dupes, function(x){which(gene.coords$gene_name == x)[-1]}))
    if(!is.null(dupes)){
      gene.coords <- gene.coords[-dupes]
    }

    # subset the DEGs:
    cur_degs <- degs %>% subset(gene %in% gene.coords$gene_name)
    gene.coords <- gene.coords %>% subset(gene_name %in% cur_degs$gene)

    # add order based on genomic coordinates:
    gene.coords$order <- 1:length(gene.coords)
    chr_n_genes <- c(chr_n_genes, length(gene.coords))

    # add the order to the deg table:
    cur_degs$order <- match(cur_degs$gene, gene.coords$gene_name)

    #logfc_thresh <- 0.25
    cur_degs$avg_log2FC <- ifelse(
      cur_degs$p_val_adj > 0.05, 0,
      cur_degs$avg_log2FC
    )

    print(dim(cur_degs))

    # plot as a heatmap:
    p <- cur_degs %>%
      ggplot(aes(x = order, y = group, fill = avg_log2FC)) +
      ggrastr::rasterise(geom_tile(), dpi=800) +
      scale_fill_gradient2(high="blue", mid="lightgrey", low="yellow", limits = c(-logfc_limit, logfc_limit)) +
      xlab(cur_chr) +
      theme(
        axis.line.x = element_blank(),
        axis.ticks.x = element_blank(),
        axis.text.x = element_blank(),
        axis.title.y = element_blank(),
        axis.line.y = element_blank(),
        axis.ticks.y = element_blank(),
        plot.margin = margin(0,0,0,0)
      )

    if(length(plot_list) > 0){
      p <- p + theme(
        axis.text.y = element_blank(),
      )
    }

    plot_list[[cur_chr]] <- p

  }


}



pdf(paste0(fig_dir, 'manheatmap_sub.pdf'), width=10, height=4)
dev.off()


# settings for function:
EnsDb <- EnsDb.Hsapiens.v86
cur_chr <- 21
cur_group <- 'MG'
size_scale = 3







gene.coords <- ensembldb::genes(EnsDb, filter = ~ gene_biotype == "protein_coding") %>% subset(seqnames == cur_chr)
'APP' %in% gene.coords$gene_name

# subset by genes in our snRNA data
gene.coords <- gene.coords[gene.coords$gene_id %in% gene_table$gene_id]
ix <- match(gene.coords$gene_id, gene_table$gene_id)
gene.coords$gene_name <- gene_table$gene_name[ix]

'APP' %in% gene.coords$gene_name

# subset the DEGs:
cur_degs <- degs %>% subset(group == cur_group & gene %in% gene.coords$gene_name)
'APP' %in% cur_degs$gene



app_id <- gene_table %>% subset(gene_name == 'APP') %>% .$gene_id



# add the DEG stats to the gene.coords object:
ix <- match(gene.coords$gene_name, cur_degs$gene)
gene.coords
gene.coords$avg_log2FC <- cur_degs[ix, 'avg_log2FC']
gene.coords$p_val_adj <- cur_degs[ix, 'p_val_adj']

# convert to dataframe
plot_df <- as.data.frame(gene.coords)

# label the top 5 genes by fold change:
# top_5 <- plot_df %>% dplyr::top_n(10, wt=avg_log2FC) %>% .$gene_name
# bottom_5 <- plot_df %>% dplyr::top_n(-5, wt=avg_log2FC) %>% .$gene_name
# label_genes <- c(top_5, bottom_5)
# plot_df$label <- ifelse(plot_df$gene_name %in% label_genes, plot_df$gene_name, '')

# label all of the significant & up-regulated genes:
label_genes <- plot_df %>% subset(avg_log2FC >= 0 & p_val_adj <= 0.05) %>% .$gene_name
plot_df$label <- ifelse(plot_df$gene_name %in% label_genes, plot_df$gene_name, '')

# color genes by p-val
plot_df$color <- ifelse(plot_df$p_val_adj <= 0.05, 'black', 'grey70')

# y-axis limit
# plot_limit <- max(abs(plot_df$avg_log2FC)) + 0.05

pval_min <- plot_df %>% subset(p_val_adj != 0) %>% .$p_val_adj %>% min
plot_df$p_val_adj <- ifelse(plot_df$p_val_adj == 0, pval_min, plot_df$p_val_adj)

# plot
p <- plot_df %>%
  ggplot(aes(x = start, y = avg_log2FC, size=-log(p_val_adj)/size_scale)) +
  geom_rect(
    data = plot_df[1,],
    aes(xmin=-Inf, xmax=Inf, ymin=-0.25, ymax=0.25), fill='grey85', alpha=0.8, color=NA
  ) +
  geom_hline(yintercept=0, linetype='dashed', color='black') +
  geom_segment(data = subset(plot_df, label != ''), aes(y=0, yend=avg_log2FC, x=start, xend=start), size=0.5, color='grey') +
  geom_point(color=plot_df$color) +
  geom_text_repel(aes(label=label), color='black', fontface='italic',  min.segment.length=0, size=3, max.overlaps=Inf) +
#  scale_y_continuous(limits=c(-1*plot_limit, plot_limit)) +
  NoLegend() +
  xlab('Chromosome 21 Coordinates')



pdf(paste0(fig_dir, 'chr21_manhat_MG.pdf'), width=8, height=4)
p
dev.off()


```
