
```{r eval=FALSE}

library(Seurat)
#library(tidyverse)
library(dplyr)
library(ggplot2)
library(cowplot)
library(Matrix)
library(viridis)
library(harmony)
library(ggpubr)
library(patchwork)
library(RColorBrewer)
library(ggrepel)
library(CellTrek)
library(hdWGCNA)

#source("/pub/smorabit/hdWGCNA/bin/spatial_functions.R")
source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')


# 2,000 MB limit:
options(future.globals.maxSize= 2000*1024^2)


setwd("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/spatial_mapping/")

fig_dir <- "figures/"
data_dir <- "data/"


# load visium & split-seq seurat objects:
seurat_vis <- readRDS('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/data/ADDS_seurat_processed.rds')

# seurat_obj <- readRDS("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/data/ADDS_integrated_scvi.rds")

# load ADDS dataset:
seurat_obj <- readRDS(file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_integrated.rds" )

# only get the FCX:
seurat_obj <- subset(seurat_obj, Tissue != 'PCC')

# saveRDS(seurat_obj, file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_integrated_FCX.rds")

# add barcode:
seurat_obj$bc <- colnames(seurat_obj)
seurat_vis$bc <- colnames(seurat_vis)

var_df <- read.delim(file = "/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_subclusters_var.tsv", sep='\t')

VariableFeatures(seurat_obj) <- var_df$X



# representative samples
human_rep_samples <- c(
  'Dec_13_2021_Human5', 'Dec_20_2021_Human1',
  'Dec_13_2021_Human6', 'Oct_2021_6',
  'Dec_13_2021_Human3', 'Dec_13_2021_Human7',
  'Nov_24_2021_VisiumHuman_12', 'Dec_13_2021_Human8'
)

#################################################################
# Load color schemes
#################################################################


human_cp <- c(
      "L1" = "#8B3D5A", "L2-3" = "#E7BDE1", "L3-4" = "#E6A4CD",
      "L3-4-5" = "#CF8BA3", "L5-6" = "#9E6D7F", "L6b" = "#CDAEB9", "WM1" = "#64BCDB", "WM2" = "#62A7D7", "WM3" = "#99C8D7")


# color scheme:
color_df <- read.csv(file='/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/cell_identity.csv')
color_df <- color_df[,2:3]

group_levels <- c(
  'EX L2', 'EX L2-3', 'EX L3-5', 'EX L5', 'EX L5-6', 'EX L6',
  'INH VIP+', 'INH', 'INH LAMP5+', 'INH PVALB+', 'INH SST+',
  'ASC1', 'ASC2', 'ASC3', 'ASC4',
  'MG1', 'MG2',
  'ODC1', 'ODC2', 'ODC3',
  'OPC1', 'OPC2', 'OPC3',
  'END Arterial', 'END Capillary',
  'T-Pericyte', 'M-Pericyte', 'SMC',
  'Perivascular Fibroblast', 'Meningeal Fibroblast'
)
color_df$group <- factor(as.character(color_df$group), levels=group_levels)
color_df <- arrange(color_df, group)
cp <- as.character(color_df$colour)
names(cp) <- as.character(color_df$group)



st_color_df <- data.frame(
  colour = as.character(human_cp),
  group = names(human_cp)
)
st_color_df$group <- factor(as.character(st_color_df$group), levels=names(human_cp))


seurat_obj$cell_identity <- factor(
  as.character(seurat_obj$cell_identity),
  levels = group_levels
)

#################################################################
# Load cluster DEGs:
#################################################################

# read combined DEGs
degs <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/cluster_markers.csv'))
degs$group <- factor(as.character(degs$group), levels=group_levels)


#################################################################
# Load celltrek coordinates:
#################################################################

anno_df <- read.csv('celltrek_coords_combined.csv')

#################################################################
# Representative Visium Samples:
#################################################################


patch <- VisDimPlot(
  seurat_vis,
  group.by = 'annotation',
  sample_labels = c("Diagnosis", "Sex", 'Age', "Sample"),
  dpi=600,
  ncol = 10,
  color_df=human_cp,
  text_size=7
)

pdf(paste0(fig_dir, 'hex_annotation_human.pdf'), width=20, height=10)
patch
dev.off()




```

Plot some examples of the co-embedding Seurat objects:
```{r eval=FALSE}

dir.create(paste0(fig_dir, 'umap_coembed'))

human_cp <- c(
      "L1" = "#8B3D5A", "L2/3" = "#E7BDE1", "L3/4" = "#E6A4CD",
      "L3/4/5" = "#CF8BA3", "L5/6" = "#9E6D7F", "L6b" = "#CDAEB9", "WM1" = "#64BCDB", "WM2" = "#62A7D7", "WM3" = "#99C8D7")


st_color_df <- data.frame(
  colour = as.character(human_cp),
  group = names(human_cp)
)
st_color_df$group <- factor(as.character(st_color_df$group), levels=names(human_cp))

file_list <- dir(data_dir)

sc_samples <- c('ADDS_humAD-97', 'ADDS_86')

coembed_files <- paste0('data/vis_', human_rep_samples, '-sc_', sc_samples,'_coembed.rds')


# for(i in 1:length(human_rep_samples)){
for(cur_file in coembed_files){

  #print(i)
  #cur_file_list <- paste0(data_dir, file_list[grepl(human_rep_samples[i], file_list)])
  #cur_file <- cur_file_list[1]

  cur_seurat <- readRDS(cur_file)


  table(cur_seurat$type)

  table(cur_seurat$annotation)
  table(cur_seurat$cell_identity)


  st_sc_group_levels <- c(group_levels, names(human_cp))
  st_sc_color_df <- rbind(color_df, st_color_df)

  cur_seurat$plot_groups <- ifelse(
    cur_seurat$type == 'st',
    as.character(cur_seurat$annotation),
    as.character(cur_seurat$cell_identity)
  )
  cur_seurat$plot_groups <- factor(as.character(cur_seurat$plot_groups), levels=st_sc_group_levels)



  plot_list <- PlotEmbedding(
    cur_seurat,
    group.by = 'plot_groups',
    split.by = 'type',
    plot_under=TRUE,
    raster_dpi = 500,
    raster_scale=0.5, point_size=1,
    color_df = st_sc_color_df,
    label=TRUE,
    plot_theme = umap_theme() + NoLegend()
  )

  outfile <- gsub('data/', '', cur_file)
  outfile <- gsub('.rds', '.pdf', outfile)

  pdf(paste0(fig_dir, 'umap_coembed/labeled_', outfile), height=5, width=10)
  print(wrap_plots(plot_list, ncol=2))
  dev.off()

}


```

plot single-cell data split by sample

```{r eval=FALSE}

plot_list <- PlotEmbedding(
  seurat_obj,
  group.by = 'cell_identity',
  split.by = 'Sample',
  plot_under=TRUE,
  raster_dpi = 500,
  raster_scale=0.5, point_size=1,
  color_df = color_df,
  label=FALSE,
  plot_theme = umap_theme() + NoLegend()
)

pdf(paste0(fig_dir, 'umap_split_samples.pdf'), height=15, width=15)
wrap_plots(plot_list[1:8], ncol=3)
dev.off()

unique(seurat_obj$Sample)[1:8]

subset(seurat_obj@meta.data, Sample == 'ADDS_humAD-97') %>% head
subset(seurat_obj@meta.data, Sample == 'ADDS_86') %>% head


```

Plot all mapped cells in representative samples

```{r eval=FALSE}

ex_neurons <- c(
  'EX L2', 'EX L2-3', 'EX L3-5', 'EX L5', 'EX L5-6', 'EX L6')

inh_neurons <- c(
  'INH VIP+', 'INH', 'INH LAMP5+', 'INH PVALB+', 'INH SST+')

glia <- c(
  'ASC1', 'ASC2', 'ASC3', 'ASC4',
  'MG1', 'MG2',
  'ODC1', 'ODC2', 'ODC3',
  'OPC1', 'OPC2', 'OPC3'
)
vasc <- c(
  'END Arterial', 'END Capillary',
  'T-Pericyte', 'M-Pericyte', 'SMC',
  'Perivascular Fibroblast', 'Meningeal Fibroblast'
)

group_list <- list(
  EX = ex_neurons,
  INH = inh_neurons,
  glia = glia,
  VASC = vasc
)

# get sc coordinates for the current visium sample
plot_list <- list()
for(i in 1:length(human_rep_samples)){

  cur_vis_sample <- human_rep_samples[i]
  print(cur_vis_sample)
  plot_list <- list()
  for(cur_group in names(group_list)){
    cur_coords <- subset(anno_df, vis_sample == cur_vis_sample & cell_identity %in% group_list[[cur_group]])

    # plot the main dataset
    p <- cur_coords %>%
      ggplot(aes(x=celltrek_1, y=celltrek_2, color=cell_identity)) +
      ggrastr::rasterise(geom_point(size=1, alpha=1), dpi=300, scale=0.5) +
      scale_color_manual(values=cp) +
      umap_theme() + coord_fixed() +
      guides(colour = guide_legend(override.aes = list(size=5))) +
      NoLegend()

    plot_list[[cur_group]] <- p
  }

  pdf(paste0(fig_dir, 'rep_samples_mapped/', human_rep_samples[i],'.pdf'), width=12, height=3)
  print(wrap_plots(plot_list, ncol=4))
  dev.off()

}

pdf(paste0(fig_dir, 'rep_samples_mapped_cell_ids.pdf'), width=18, height=3)
print(wrap_plots(plot_list, ncol=8))
dev.off()


```


Plot expression of marker genes in mapped cells

```{r eval=FALSE}

ex_neurons <- c(
  'EX L2', 'EX L2-3', 'EX L3-5', 'EX L5', 'EX L5-6', 'EX L6')

inh_neurons <- c(
  'INH VIP+', 'INH', 'INH LAMP5+', 'INH PVALB+', 'INH SST+')

glia <- c(
  'ASC1', 'ASC2', 'ASC3', 'ASC4',
  'MG1', 'MG2',
  'ODC1', 'ODC2', 'ODC3',
  'OPC1', 'OPC2', 'OPC3'
)
vasc <- c(
  'END Arterial', 'END Capillary',
  'T-Pericyte', 'M-Pericyte', 'SMC',
  'Perivascular Fibroblast', 'Meningeal Fibroblast'
)

group_list <- list(
  EX = ex_neurons,
  INH = inh_neurons,
  glia = glia,
  VASC = vasc
)

cur_vis_sample <- 'Dec_13_2021_Human8'
cur_vis_sample <- 'Dec_13_2021_Human5'

layer_markers <- list(
  'Layer 1' = c('C4or31', 'RELN', 'INPP4B'),
  'Layer 1/2' = c('CHRNA7', 'CNR1', 'CXCL14'),
  'Layer 2' = c('PVRL3', 'RASGRF2', 'WFS1', 'C1QL2', 'CARTPT'),
  'Layer 2/3' = c('GSG1L', 'IGSF11', 'KCNIP2', 'C20orf103', 'CALB1'),
  'Layer 3' = c('PRSS12', 'MFGE8', 'SV2C'),
  'Layer 4' = c('RORB', 'CACNG5', 'CHRNA3', 'KCNIP1', 'PDYN'),
  'Layer 5' = c('VAT1L', 'HTR2C'),
  'Layer 5/6' = c('CPNE7', 'ETV1', 'TOX', 'B3GALT2', 'KCNK2', 'PCP4', 'PDE1A', 'RPRM', 'PCDH20'),
  'Layer 6' = c('FOXP2', 'NTNG2', 'SYT10', 'SYT6', 'TH', 'TMEM163', 'AKR1C2', 'AKR1C3', 'NPY2R', 'OPRK1', 'SEMA3C', 'SYNPR'),
  'Layer 6/6B' = c('GABRA5', 'TLE4', 'PCDH17'),
  'Layer 6B/WM' = c('ADRA2A', 'CTGF'),
  'WM' = c('MBP', 'MOG', 'PLP1')
)

cur_coords <- subset(anno_df, vis_sample == cur_vis_sample)
cur_seurat <- seurat_obj[,cur_coords$sc_bc]
rownames(cur_coords) <- cur_coords$sc_bc
cur_coords <- cur_coords[colnames(cur_seurat),]

cur_seurat@reductions$spatial <- CreateDimReducObject(
  embeddings = as.matrix(cur_coords[,c('celltrek_1', 'celltrek_2')])
)

for(cur_gene in unlist(layer_markers)){

  print(cur_gene)

  if(!(cur_gene %in% rownames(cur_seurat))){
    next
  }
  if(!(cur_gene %in% rownames(seurat_vis))){
    next
  }
 
  p1 <- FeatureEmbedding(
    cur_seurat,
    features = cur_gene,
    reduction = 'spatial',
    point_size=1,
    plot_min = 'q0',
    plot_max = 'q100'
  )

  p2 <- SampleFeaturePlot(
          seurat_vis,
          feature=cur_gene,
          samples_to_plot = cur_vis_sample,
          sample_labels = c("Diagnosis", "Sex", 'Age'),
          ncol = 1,
          raster=TRUE,
          plot_max = 'q100',
          plot_min = 'q0',
          colfunc = inferno,
          rev_colors=TRUE,
          dpi=400,
        )


  pdf(paste0(fig_dir, 'spatial_featureplots/',cur_vis_sample, '-', cur_gene,'.pdf'), width=8, height=4)
  print(p2 | p1)
  dev.off()

}





# get sc coordinates for the current visium sample
plot_list <- list()
for(i in 1:length(human_rep_samples)){

  cur_vis_sample <- human_rep_samples[i]
  print(cur_vis_sample)
  plot_list <- list()
  for(cur_group in names(group_list)){
    cur_coords <- subset(anno_df, vis_sample == cur_vis_sample & cell_identity %in% group_list[[cur_group]])

    # plot the main dataset
    p <- cur_coords %>%
      ggplot(aes(x=celltrek_1, y=celltrek_2, color=cell_identity)) +
      ggrastr::rasterise(geom_point(size=1, alpha=1), dpi=300, scale=0.5) +
      scale_color_manual(values=cp) +
      umap_theme() + coord_fixed() +
      guides(colour = guide_legend(override.aes = list(size=5))) +
      NoLegend()

    plot_list[[cur_group]] <- p
  }

  pdf(paste0(fig_dir, 'rep_samples_mapped/', human_rep_samples[i],'.pdf'), width=12, height=3)
  print(wrap_plots(plot_list, ncol=4))
  dev.off()

}

pdf(paste0(fig_dir, 'rep_samples_mapped_cell_ids.pdf'), width=18, height=3)
print(wrap_plots(plot_list, ncol=8))
dev.off()


```




Plot density of mapped cells in each cluster

```{r eval=FALSE}

# marker gene plotting
plot_markers <- FALSE
n_degs <- 5

# how many columns in the plot?
n_columns <- 3

cur_celltype <- 'INH'
cur_celltype <- c('OPC', 'ODC')
cur_celltype <- c('FBR', 'PER', 'SMC', 'END')
cur_celltype <- 'ASC'
cur_celltype <- 'EX'
cur_celltype <- 'MG'

# get sc coordinates for the current visium sample
for(i in 1:length(human_rep_samples)){

  cur_vis_sample <- human_rep_samples[i]
  cur_coords <- subset(anno_df, vis_sample == cur_vis_sample)

  clusters <- cur_coords %>% subset(cell_type %in% cur_celltype) %>% .$cell_identity %>% as.character %>% unique
  clusters <- clusters[order(clusters)]

  # set width and height by number of clusters:
  width <- 8
  height <- 3 * ceiling(length(clusters)/n_columns)

  plot_list <- list()
  for(cur_cluster in clusters){

    print(cur_cluster)

    # plot the main dataset
    p <- cur_coords %>% subset(cell_identity == cur_cluster) %>%
      ggplot(aes(x=celltrek_1, y=celltrek_2, color=cell_identity)) +
      ggrastr::rasterise(
        geom_point(data=cur_coords, size=0.5, color='lightgrey'), dpi=200, scale=1
      ) +
      ggrastr::rasterise(geom_point(size=0.75, alpha=1), dpi=400, scale=1) +
      #metR::geom_contour_tanaka()
      geom_density_2d(color='black')

    # add plotting options
    p <- p +
      scale_color_manual(values=cp) +
      umap_theme() + coord_fixed() +
      guides(colour = guide_legend(override.aes = list(size=5))) +
      NoLegend() +
      ggtitle(cur_cluster) +
      theme(
        plot.title = element_text(hjust=0.5, face='bold', color=as.character(cp[cur_cluster]))
      )

    plot_list[[cur_cluster]] <- p


    ################################################################################
    # cluster marker sc feature plots
    ################################################################################

    if(plot_markers){

      cur_degs <- subset(degs, group == cur_cluster) %>%
        slice_max(n=n_degs, order_by=avg_log2FC) %>% .$gene

      # plot these genes:
      cur_gene <- cur_degs[1]
      cur_exp <- FetchData(seurat_obj, cur_gene, slot='data')

      ix <- match(cur_coords$sc_bc, rownames(cur_exp))
      cur_coords$expression <- as.numeric(cur_exp[ix,1])

      for(cur_gene in cur_degs){

        # plot the main dataset
        p <- cur_coords %>%
          subset(expression != 0) %>%
          arrange(expression) %>%
          ggplot(aes(x=celltrek_1, y=celltrek_2, color=expression)) +
          ggrastr::rasterise(
            geom_point(data=cur_coords, color='grey', size=0.5, alpha=0.05), dpi=100, scale=0.5
          ) +
          ggrastr::rasterise(geom_point(size=0.75), dpi=600, scale=1)

        # add plotting options
        p <- p +
          scale_color_gradientn(colors=rev(inferno(256))) +
          umap_theme() + coord_fixed() +
          ggtitle(cur_gene) +
          theme(
            plot.title = element_text(hjust=0.5)
          )


        p_vis <- SampleFeaturePlot(
          seurat_vis,
          feature=cur_gene,
          samples_to_plot = cur_vis_sample,
          sample_labels = c("Diagnosis", "Sex", 'Age'),
          ncol = 1,
          raster=TRUE,
          plot_max = 'q99',
          plot_min = 'q0',
          colfunc = inferno,
          rev_colors=TRUE,
          dpi=400,
        )

        pdf(paste0(fig_dir, 'DEGs/', cur_cluster, '_', cur_gene, '_',  cur_vis_sample,  '.pdf'), width=8, height=4)
        print(p + p_vis)
        dev.off()

      }
    }
  }

  pdf(paste0(fig_dir, 'density/', cur_celltype, '_', cur_vis_sample, '_cell_ids.pdf'), width=width, height=height)
  print(wrap_plots(plot_list, ncol=n_columns))
  dev.off()

}

```


Proportion bar charts of cells mapped to regions in each spatial sample

```{r eval=FALSE}



################################################################################
# snRNA-seq Proportion of Condition
################################################################################

cur_coords <- subset(cur_coords, annotation %in% seurat_vis$annotation)
cur_coords$annotation <- factor(
  as.character(cur_coords$annotation),
  levels = levels(seurat_vis$annotation)
)
cur_coords$cell_identity <- factor(
  as.character(cur_coords$cell_identity),
  levels = levels(seurat_obj$cell_identity)
)

batches <- cur_coords$annotation %>% unique %>% as.character
cur_clusters  <- cur_coords$cell_identity %>% unique %>% as.character

scale_vec <- table(cur_coords$annotation)

proportion_df <- data.frame()
for(i in 1:length(cur_clusters)){
  cluster_meta <- subset(cur_coords, cell_identity == cur_clusters[i] )
  cur_df <- as.data.frame(table(cluster_meta$annotation)) %>% dplyr::rename(Count = Freq)

  # compute the non-normalized % of cells in each group
  cur_df$Freq <- cur_df$Count / sum(cur_df$Count)

  # scale frequency to the total number of clusters in each sample
  cur_df$Scaled <- cur_df$Count / scale_vec
  cur_df$Scaled <- cur_df$Scaled / sum(cur_df$Scaled)

  # add to ongoing proportion df
  cur_df$cluster <- cur_clusters[i]
  proportion_df <- rbind(proportion_df, cur_df)
}

proportion_df <- dplyr::rename(proportion_df, annotation = Var1)
proportion_df$cluster <- factor(proportion_df$cluster, levels=levels(cur_coords$cell_identity))

p <- ggplot(proportion_df, aes(y=Scaled, x=cluster, fill=annotation)) +
  geom_bar(stat='identity') +
#  geom_hline(yintercept = 0.5, linetype='dashed') +
  scale_y_continuous(expand = c(0,0)) +
  scale_fill_manual(values=human_cp) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    axis.text.x = element_text(angle=45, hjust=1),
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    axis.line.y = element_line(colour = "black"),
    axis.line.x = element_blank(),
    #panel.border = element_rect(linewidth=1, fill=NA, color='black')
  )

pdf(paste0(fig_dir, 'test_cluster_proportion_region_scaled2.pdf'), height=4, width=10)
print(p)
dev.off()

p <- ggplot(proportion_df, aes(y=Freq, x=cluster, fill=Sample)) +
  geom_bar(stat='identity') +
  geom_hline(yintercept = 0.5, linetype='dashed') +
  scale_y_continuous(expand = c(0,0)) +
  #scale_fill_manual(values=c('gray65', 'gray45')) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    axis.text.x = element_text(angle=45, hjust=1),
    axis.title.x = element_blank(),
    legend.title = element_blank(),
    axis.line.y = element_line(colour = "black"),
    axis.line.x = element_blank()
  )

pdf(paste0(fig_dir, 'cluster_proportion_sample_raw.pdf'), height=4, width=10)
print(p)
dev.off()




```


```{r eval=FALSE}

anno_df_full <- anno_df
annotation_names <- c('WM', as.character(unique(seurat_vis$annotation)))

# re-name white matter:
anno_df$anno_fix <- ifelse(
  grepl('WM', anno_df$annotation) & !(grepl('L', anno_df$annotation)),
  'WM',
  anno_df$annotation
)


anno_df <- subset(anno_df_full, anno_fix %in% annotation_names)
anno_df$sc_bc <- as.character(anno_df$sc_bc)
anno_df$anno_fix <- factor(
  as.character(anno_df$anno_fix),
  levels = c('L1', 'L2/3', 'L3/4', 'L3/4/5', 'L5/6', 'L6b', 'WM')
)

# count how many times each cell was mapped to each annotation
test <- anno_df %>% dplyr::group_by(sc_bc, anno_fix, .drop=FALSE) %>%
    dplyr::summarize(Freq=n())

bc_counts <- table(anno_df$sc_bc)
ix <- match(test$sc_bc, names(bc_counts))
test$Freq / as.numeric(bc_counts[ix])
test$prop <- test$Freq / as.numeric(bc_counts[ix])


props <- data.frame(bc = colnames(seurat_obj))
for(cur_group in as.character(unique(test$anno_fix))){

  test_prop <- subset(test, anno_fix == cur_group)
  ix <- match(colnames(seurat_obj), as.character(test_prop$sc_bc))
  cur_prop <- data.frame(
    cur_group = test_prop$prop[ix]
  )
  names(cur_prop) <- cur_group
  props <- cbind(props, cur_prop)

#  p <- VlnPlot(seurat_plot, features='prop', group.by='cell_identity', pt.size=0) + NoLegend()


}

meta <- seurat_obj@meta.data
seurat_obj@meta.data <- cbind(meta, props[,-1])

clusters <- unique(seurat_obj$cell_identity)
tmp <- levels(seurat_obj$cell_identity)
clusters <- tmp[tmp %in% clusters]

p <- custom_vln(
    seurat_obj,
    features = names(props[,-1]),
    group.by = 'cell_identity',
    groups = clusters,
    add_boxplot=FALSE,
    group_color_df = color_df
  )




pdf(paste0(fig_dir, 'test_prop_vln_bigcolor.pdf'), width=8, height=5)
p 
dev.off()

pdf(paste0(fig_dir, 'test_prop_vln.pdf'), width=10, height=4)
p
dev.off()





mg_bcs <- colnames(seurat_plot[, seurat_plot$cell_type == 'MG'])
test_mg <- test %>% subset(sc_bc %in% mg_bcs)

test_mg %>% subset(anno_fix == 'WM' & Freq > 20) %>% dim


test <- test %>% group_by(sc_bc) %>% top_n(1, wt=Freq) %>%
  mutate(anno_fix = paste0(anno_fix, collapse=' ')) %>%
  distinct()


sum(test$anno_fix %in% annotation_names) / nrow(test)

table(test$anno_fix)


test2 <- subset(test, anno_fix %in% annotation_names)



seurat_plot <- seurat_obj[,test2$sc_bc]
ix <- match(as.character(colnames(seurat_plot)), test2$sc_bc)
seurat_plot$spatial_anno <- as.character(test2$anno_fix)[ix]



p <- PlotEmbedding(
  seurat_plot,
  group.by = 'spatial_anno',
  raster_dpi = 600,
  raster_scale=0.5, point_size=1,
  plot_theme = umap_theme()
)


p1 <- DimPlot(seurat_plot, group.by = 'spatial_anno', raster=F, label=TRUE) + umap_theme()
p2 <- DimPlot(seurat_plot, group.by = 'cell_identity', raster=F, label=TRUE) + umap_theme()

pdf(paste0(fig_dir, 'umap_spatial_anno.pdf'), width=6, height=6)
p
dev.off()


pdf(paste0(fig_dir, 'umap_spatial_anno.pdf'), width=12, height=4)
p1 + p2
dev.off()

test3 <- subset(seurat_plot@meta.data, cell_type == 'MG')
table(test3$spatial_anno) / nrow(test3) * 100

```
