
Load ADDS Seurat data:

```{r eval=FALSE}
# conda activate cicero

library(Seurat)
library(tidyverse)
library(cowplot)
library(Matrix)
library(viridis)
library(harmony)
library(ggpubr)
library(patchwork)
library(RColorBrewer)
library(ggrepel)
library(tictoc)


theme_set(theme_cowplot())
colfunc <- colorRampPalette(rev(brewer.pal(9, 'Purples' )[2:9]))
setwd("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/")

# directories
data_dir <- "data/"
fig_dir <- 'figures/'

umap_theme <- theme(
  axis.line=element_blank(),
  axis.text.x=element_blank(),
  axis.text.y=element_blank(),
  axis.ticks=element_blank(),
  axis.title.x=element_blank(),
  axis.title.y=element_blank(),
  panel.background=element_blank(),
  panel.border=element_blank(),
  panel.grid.major=element_blank(),
  panel.grid.minor=element_blank(),
  plot.background=element_blank(),
  plot.title=element_text(hjust=0.5)
)

# re-load integrated
# seurat_obj <- readRDS(file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_AD_integrated.rds" )

seurat_adds <- readRDS(file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_integrated.rds" )

tic(); dim(subset(seurat_adds, cell_type == 'ASC')); toc()

# re-load ADDS seurat object:
seurat_AD <- readRDS(file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/AD_integrated.rds" )

var_df <- read.delim(file = "/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/data/ADDS_subclusters_var.tsv", sep='\t')

VariableFeatures(seurat_adds) <- var_df$X
VariableFeatures(seurat_AD) <- var_df$X
VariableFeatures(seurat_obj) <- var_df$X

tmp <- seurat_obj@meta.data %>% subset(Study == "Zhou_2020")

table(tmp$Sample, tmp$Diagnosis)

# add colors
color_df <- read.csv(file=paste0(data_dir, 'cell_identity.csv'))
seurat_obj@misc$color_scheme <- color_df
human_cp <- color_df$colour
names(human_cp) <- color_df$group

# load gene name table from split-pipe:
sp_gene_table <- read.csv("/dfs7/swaruplab/smorabit/data/ADDS_2021/analysis/Batch1/Sublibrary1_S1/all-well/DGE_filtered/genes.csv")

# load gene name table from kallisto
kb_gene_table <- read.table("/dfs7/swaruplab/shared_lab/cross-disorder/count_matrices/AD_Mathys_2019/D17-8777/counts_unfiltered/genes.tsv")
kb_gene_table <- kb_gene_table[,1:2]
colnames(kb_gene_table) <- c('gene_id', 'gene_name')
kb_gene_table$gene_id <- do.call(rbind, strsplit(kb_gene_table$gene_id, '[.]'))[,1]

library(EnsDb.Hsapiens.v86)
library(GenomicRanges)
library(ensembldb)

cp <- c("Control" = "#B8DBC5", "earlyAD" = "#E7BDE1" , "AD" = "#CF8BA3", "AD_DS" = "#9E6D7F")


```

ADDS samples in each age group

```{r eval=FALSE}

meta <- subset(seurat_adds@meta.data, Diagnosis == 'DSAD')

meta %>% dplyr::select(c(Sample, Age, Sex, Region)) %>% distinct %>% arrange(Age) %>% write.csv('data/sample_ages.csv')

```


Combine individual DEG tables into one big table:

```{r eval=FALSE}

# DEGs for clusters
test_dir <- "/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/condition/"

for(cur_test in dir(test_dir)){
  print(cur_test)
  DEG_dir <- paste0("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/condition/", cur_test, '/')
  DEG_tests <- dir(DEG_dir)

  # combine  all tests into one table:
  combined <- Reduce(rbind, lapply(dir(DEG_dir), function(file){
    read.csv(paste0(DEG_dir, file))
  }))


  # write full table
  write.csv(combined, file=paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/', cur_test, '_degs.csv'), quote=FALSE, row.names=FALSE)

}


```

Check gene for vivek

```{r eval=FALSE}


source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')

group_levels <- c(
  'EX L2', 'EX L2-3', 'EX L3-5', 'EX L5', 'EX L5-6', 'EX L6',
  'INH VIP+', 'INH', 'INH LAMP5+', 'INH PVALB+', 'INH SST+',
  'ODC1', 'ODC2', 'ODC3',
  'OPC1', 'OPC2', 'OPC3',
  'MG1', 'MG2',
  'ASC1', 'ASC2', 'ASC3', 'ASC4',
  'END Arterial', 'END Capillary',
  'T-Pericyte', 'M-Pericyte', 'SMC',
  'Perivascular Fibroblast', 'Meningeal Fibroblast'
)
group_groups <- c(rep("EX", 6), rep("INH", 5), rep("ODC", 6), rep("GLIA", 6), rep("VASC", 7))
group_df <- data.frame(
  cell_id = group_levels,
  group = group_groups
)

#  cur_test <- 'DS_ct_PCC'; gene_table <- sp_gene_table
cur_test <- 'DS_ci_PCC'; gene_table <- sp_gene_table
cur_test <- 'DS_ci_FCX'; gene_table <- sp_gene_table
cur_test <- 'AD_ci'; gene_table <- kb_gene_table

# load DSAD vs Control DEGs:
degs <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/', cur_test, '_degs.csv'))
degs$group <- factor(as.character(degs$group), levels=rev(group_levels))

subset(degs, gene == 'IL21R' & p_val_adj < 0.05)




```

DEG heatmap ordered by chromosome:

```{r eval=FALSE}

source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')

group_levels <- c(
  'EX L2', 'EX L2-3', 'EX L3-5', 'EX L5', 'EX L5-6', 'EX L6',
  'INH VIP+', 'INH', 'INH LAMP5+', 'INH PVALB+', 'INH SST+',
  'ODC1', 'ODC2', 'ODC3',
  'OPC1', 'OPC2', 'OPC3',
  'MG1', 'MG2',
  'ASC1', 'ASC2', 'ASC3', 'ASC4',
  'END Arterial', 'END Capillary',
  'T-Pericyte', 'M-Pericyte', 'SMC',
  'Perivascular Fibroblast', 'Meningeal Fibroblast'
)
group_groups <- c(rep("EX", 6), rep("INH", 5), rep("ODC", 6), rep("GLIA", 6), rep("VASC", 7))
group_df <- data.frame(
  cell_id = group_levels,
  group = group_groups
)

#  cur_test <- 'DS_ct_PCC'; gene_table <- sp_gene_table
cur_test <- 'DS_ci_PCC'; gene_table <- sp_gene_table
cur_test <- 'DS_ci_FCX'; gene_table <- sp_gene_table
cur_test <- 'AD_ci'; gene_table <- kb_gene_table

# load DSAD vs Control DEGs:
degs <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/', cur_test, '_degs.csv'))
degs$group <- factor(as.character(degs$group), levels=rev(group_levels))

logfc_thresh <- 0.5
#logfc_thresh <- 0.25

# plot chromosome heatmap
color_df <- seurat_obj@misc$color_scheme
color_df$group <- factor(
  as.character(color_df$group),
    levels=levels(degs$group)
)
color_df <- arrange(color_df, group)

p <- PlotDEGsChromosome(
  degs,
  gene_table = gene_table,
  EnsDb = EnsDb.Hsapiens.v86,
  chr_names = c(as.character(1:22), 'X'),
  raster_dpi=800,
  logfc_thresh = logfc_thresh,
  plot_limit=1,
  color_df = color_df
)


pdf(paste0(fig_dir, 'manheatmap_', cur_test, '_', as.character(logfc_thresh), 'FC.pdf'), width=12, height=5)
p
dev.off()


################################################################################
# plot the number of DEGs per chromosome
################################################################################

chr_names <- c(as.character(1:22), 'X')

chr_degs <- PlotDEGsChromosome(
  degs,
  gene_table = gene_table,
  EnsDb = EnsDb.Hsapiens.v86,
  chr_names = chr_names,
  raster_dpi=800,
  logfc_thresh = 0.25,
  plot_limit=1,
  color_df = color_df,
  return_table=TRUE
)

# set factor levels
chr_degs$chr <- factor(
  as.character(chr_degs$chr),
  levels = chr_names
)

# get the number of genes on each chromosome:
n_genes_chr <- table(seqnames(ensembldb::genes(EnsDb.Hsapiens.v86, filter = ~ gene_biotype == "protein_coding")))
n_genes_chr <- ngenes_chr[chr_names]

# up-regulated:
chr_up <- subset(chr_degs, avg_log2FC > 0.25 & p_val_adj < 0.05)
n_degs_up <- table(chr_up$group, chr_up$chr)
percent_degs_up <- t(apply(n_degs_up, 1, '/', n_genes_chr))
df_up <- reshape2::melt(n_degs_up) %>% dplyr::rename(c(group = Var1, chr = Var2))
df_up$direction <- 1

# down-regulated:
chr_down <- subset(chr_degs, avg_log2FC < -0.25 & p_val_adj < 0.05)
n_degs_down <- table(chr_down$group, chr_down$chr)
percent_degs_down <- t(apply(n_degs_down, 1, '/', n_genes_chr))
df_down <- reshape2::melt(n_degs_down) %>% dplyr::rename(c(group = Var1, chr = Var2))
df_down$direction <- -1

plot_df <- rbind(df_up, df_down)
plot_df$value <- plot_df$value * plot_df$direction

# try making a plot:
p <- plot_df %>%
  ggplot(aes(fill = group, y = value, x = chr)) +
  geom_bar(position='stack', stat='identity') +
  geom_hline(yintercept=0, color='black') +
  scale_fill_manual(values=human_cp) +
  xlab('Chromosome') +
  ylab(expression(N[DEGs]))

    
pdf(paste0(fig_dir, 'ndegs_chromosome_', cur_test, '_snRNA.pdf'), width=8, height=3)
p + NoLegend()
dev.off()


```

Correlation between logFC in AD & ADDS

```{r eval=FALSE}

library(RRHO)
library(viridis)
library(ggpubr)

# load DSAD vs Control DEGs:
degs_ad <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/AD_ci_degs.csv'))
degs_adds <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/DS_ci_FCX_degs.csv'))

# add gene_id column to both:
ix <- match(degs_ad$gene, kb_gene_table$gene_name)
degs_ad$gene_id <- kb_gene_table$gene_id[ix]

ix <- match(degs_adds$gene, sp_gene_table$gene_name)
degs_adds$gene_id <- sp_gene_table$gene_id[ix]

# genes in both:
genes.keep <- intersect(kb_gene_table$gene_id, sp_gene_table$gene_id)
degs_ad <- subset(degs_ad, gene_id %in% genes.keep)
degs_adds <- subset(degs_adds, gene_id %in% genes.keep)

# plot settings
groups <- unique(degs_ad$group)
plot_list <- list()
rrho_plot_list <- list()
cor_list <- c()
signif_only <- TRUE
run_rrho <- FALSE
rrho_maxval <- 750

# run loop over each group
# initialize progress bar
pb <- utils::txtProgressBar(min = 0, max = length(groups), style = 3, width = 50, char = "=")
counter <- 1


for(cur_group in groups){
  print(cur_group)
  setTxtProgressBar(pb, counter)


  cur_degs_ad <- subset(degs_ad, group == cur_group)
  cur_degs_adds <- subset(degs_adds, group == cur_group)

  genes.keep <- intersect(cur_degs_ad$gene_id, cur_degs_adds$gene_id)
  cur_degs_ad <- subset(cur_degs_ad, gene_id %in% genes.keep)
  cur_degs_adds <- subset(cur_degs_adds, gene_id %in% genes.keep)

  # make sure they are in the same order:
  rownames(cur_degs_ad) <- cur_degs_ad$gene_id
  rownames(cur_degs_adds) <- cur_degs_adds$gene_id
  cur_degs_adds <- cur_degs_adds[cur_degs_ad$gene_id,]

  # join the two dataframes
  plot_df <- dplyr::inner_join(cur_degs_ad, cur_degs_adds, by = 'gene_id')

  logfc_thresh <- 0.05
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Consistent", "")
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Consistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Inconsistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Inconsistent", plot_df$group)
  group_colors <- c('grey', 'blue', 'yellow')

  if(signif_only){
    plot_df <- subset(plot_df, p_val_adj.x < 0.05 | p_val_adj.y < 0.05)
    if(nrow(plot_df) == 0){next}
  }


  #plot_df <- plot_df %>% subset(group != '')
  #group_colors <- c('blue', 'yellow')

  cur_cor <-  cor(x=as.numeric(plot_df$avg_log2FC.x), y=as.numeric(plot_df$avg_log2FC.y))
  cor_list <- c(cor_list, cur_cor)
  print(table(plot_df$group))
  print(dim(plot_df))

  plot_range <- max(max(plot_df$avg_log2FC.x), max(plot_df$avg_log2FC.y))
  p <- plot_df %>%
    ggplot(aes(x = avg_log2FC.x, y = avg_log2FC.y, color=group)) +
    geom_hline(yintercept = 0, linetype='dashed', color='grey') +
    geom_vline(xintercept = 0, linetype='dashed', color='grey') +
    ggrastr::rasterise(geom_point(), dpi=500, scale=0.5) +
    scale_color_manual(values=group_colors) +
    geom_smooth(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='lm', color='black') +
    stat_cor(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='pearson', size=2) +
    xlim(c(-plot_range, plot_range)) +
    ylim(c(-plot_range, plot_range)) +
    theme(
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      axis.text.x = element_text(size=6),
      axis.text.y = element_text(size=6),
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      plot.title = element_text(hjust=0.5, size=7, vjust=-1)
    ) +
    coord_fixed(ratio=1) + NoLegend() +
    xlab(bquote("AD Average log"[2]~"(Fold Change)")) +
    ylab(bquote("ADDS Average log"[2]~"(Fold Change)")) +
    ggtitle(cur_group)
    #annotate("text", label=paste0("N=",sum(plot_df$group == 'Consistent')), x=-plot_range + 0.2, y=-plot_range+0.2, color='blue')

  plot_list[[cur_group]] <- p

  if(run_rrho){


    # set up gene lists
    gl1 <- plot_df[,c('gene_id', 'avg_log2FC.x')]
    gl2 <- plot_df[,c('gene_id', 'avg_log2FC.y')]

    # run rrho
    test <- RRHO(gl1, gl2, alternative='enrichment', BY=TRUE)
    overlap_df <- reshape2::melt(test$hypermat.by)
    overlap_df$value <- ifelse(overlap_df$value > rrho_maxval, rrho_maxval, overlap_df$value)

    # plot rrho heatmap
    p <- ggplot(overlap_df, aes(x=Var1, y=Var2, fill=value, color=value)) +
      ggrastr::rasterise(geom_tile(), dpi=500) +
      scale_fill_gradientn(colors=magma(256), limits=c(0, rrho_maxval)) +
      scale_color_gradientn(colors=magma(256), limits=c(0, rrho_maxval)) +
      theme(
        plot.title=element_text(hjust=0.5),
        axis.line=element_blank(),
        axis.ticks=element_blank(),
        axis.text.x = element_blank(),
        axis.text.y = element_blank(),
        axis.title = element_blank(),
        plot.margin=margin(0,0,0,0)
      ) + coord_equal() + ggtitle(cur_group)

    rrho_plot_list[[cur_group]] <- p

  }

  # update progress bar
  counter <- counter+1
}

# close progress bar
close(pb)


names(cor_list) <- names(plot_list)
plot_list <- plot_list[rev(order(cor_list))]
rrho_plot_list <- rrho_plot_list[rev(order(cor_list))]

plot_list <- lapply(plot_list, function(x){
  x + theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    plot.margin = margin(0,0,0,0),
    plot.title = element_text(vjust=-1)
  )
})

pdf(paste0(fig_dir, 'deg_corr_combined_signif2.pdf'), width=4.5, height=9)
wrap_plots(plot_list, ncol=4)
dev.off()

pdf(paste0(fig_dir, 'deg_rrho_combined.pdf'), width=12, height=10)
wrap_plots(rrho_plot_list, ncol=6) + plot_layout(guides='collect')
dev.off()

```


Correlation between logFC in ADDS FCX & PCC

```{r eval=FALSE}

library(viridis)

# load DSAD vs Control DEGs:
degs_ad <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/DS_ci_PCC_degs.csv'))
degs_adds <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/DS_ci_FCX_degs.csv'))

# plot settings
groups <- unique(degs_ad$group)
plot_list <- list()
rrho_plot_list <- list()
cor_list <- c()
signif_only <- TRUE

# run loop over each group
# initialize progress bar
pb <- utils::txtProgressBar(min = 0, max = length(groups), style = 3, width = 50, char = "=")
counter <- 1

degs_ad %>% subset(p_val_adj < 0.05) %>% .$avg_log2FC %>% abs %>% min

for(cur_group in groups){
  print(cur_group)
  setTxtProgressBar(pb, counter)


  cur_degs_ad <- subset(degs_ad, group == cur_group)
  cur_degs_adds <- subset(degs_adds, group == cur_group)

  # make sure they are in the same order:
  rownames(cur_degs_ad) <- cur_degs_ad$gene
  rownames(cur_degs_adds) <- cur_degs_adds$gene
  cur_degs_adds <- cur_degs_adds[cur_degs_ad$gene,]

  # join the two dataframes
  plot_df <- dplyr::inner_join(cur_degs_ad, cur_degs_adds, by = 'gene')
  library(ggpubr)
  p1 <- ggpubr::ggscatter(
    plot_df,
    x = 'avg_log2FC.x',
    y = 'avg_log2FC.y',
    add = 'reg.line',
    add.params = list(color='blue', fill='lightgrey'),
  ) + stat_cor(method='pearson')


  logfc_thresh <- 0.05
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Consistent", "")
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Consistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Inconsistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Inconsistent", plot_df$group)
  group_colors <- c('grey', 'blue', 'yellow')

  if(signif_only){
    plot_df <- subset(plot_df, p_val_adj.x < 0.05 | p_val_adj.y < 0.05)
    if(nrow(plot_df) == 0){next}
  }


  # plot_df <- plot_df %>% subset(group != '')
  # group_colors <- c('blue', 'yellow')

  cur_cor <-  cor(x=as.numeric(plot_df$avg_log2FC.x), y=as.numeric(plot_df$avg_log2FC.y))
  cor_list <- c(cor_list, cur_cor)
  print(table(plot_df$group))
  print(dim(plot_df))

  plot_range <- max(max(plot_df$avg_log2FC.x), max(plot_df$avg_log2FC.y))
  p <- plot_df %>%
    ggplot(aes(x = avg_log2FC.x, y = avg_log2FC.y, color=group)) +
    geom_hline(yintercept = 0, linetype='dashed', color='grey') +
    geom_vline(xintercept = 0, linetype='dashed', color='grey') +
    ggrastr::rasterise(geom_point(), dpi=500) +
    scale_color_manual(values=group_colors) +
    geom_smooth(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='lm', color='black') +
    stat_cor(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='pearson') +
    xlim(c(-plot_range, plot_range)) +
    ylim(c(-plot_range, plot_range)) +
    theme(
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      plot.title = element_text(hjust=0.5)
    ) +
    coord_fixed(ratio=1) + NoLegend() +
    xlab(bquote("PCC Average log"[2]~"(Fold Change)")) +
    ylab(bquote("FCX Average log"[2]~"(Fold Change)")) +
    ggtitle(cur_group)
    #annotate("text", label=paste0("N=",sum(plot_df$group == 'Consistent')), x=-plot_range + 0.2, y=-plot_range+0.2, color='blue')

  plot_list[[cur_group]] <- p

  # update progress bar
  counter <- counter+1
}

# close progress bar
close(pb)


names(cor_list) <- names(plot_list)
plot_list <- plot_list[rev(order(cor_list))]

plot_list <- lapply(plot_list, function(x){
  x + theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    plot.margin = margin(0,0,0,0),
    plot.title = element_text(vjust=-0.2)
  )
})


pdf(paste0(fig_dir, 'deg_corr_FCX_PCC_combined_signif.pdf'), width=15, height=18)
wrap_plots(plot_list, ncol=5)
dev.off()


```


Volcano plots:

```{r eval=FALSE}

library(ggrepel)

cp <- c("Control" = "#B8DBC5", "earlyAD" = "#E7BDE1" , "AD" = "#CF8BA3", "AD_DS" = "#9E6D7F")

cur_test <- 'DS_ci_PCC'; color1 <- cp['AD_DS']; color2 <- cp['Control']
cur_test <- 'DS_ci_FCX'; color1 <- cp['AD_DS']; color2 <- cp['Control']
cur_test <- 'AD_ci'; color1 <- cp['AD']; color2 <- cp['Control']


# load DSAD vs Control DEGs:
degs <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/', cur_test, '_degs.csv'))
degs$group <- factor(as.character(degs$group), levels=rev(group_levels))
name <- cur_test

# exclude MT genes
degs <- degs[!grepl('MT-', degs$gene),]

# lowest non-zero value
lowest <- degs %>% subset(p_val_adj != 0) %>% top_n(-1, wt=p_val_adj) %>% .$p_val_adj
degs$p_val_adj <- ifelse(degs$p_val_adj == 0, lowest, degs$p_val_adj)

nlabel <- 5

# label the top and bottom significant genes by log fold change
cur_degs <- Reduce(rbind, lapply(unique(degs$group), function(x){
  cur <- subset(degs, group == x)

  top_thresh <- cur %>% subset(p_val_adj <= 0.05 & avg_log2FC > 0) %>% top_n(nlabel, wt=avg_log2FC) %>% .$avg_log2FC %>% min
  bottom_thresh <- cur %>% subset(p_val_adj <= 0.05 & avg_log2FC < 0) %>% top_n(-1*nlabel, wt=avg_log2FC) %>% .$avg_log2FC %>% max

  cur$anno <- ifelse(cur$p_val_adj <= 0.05 & cur$avg_log2FC >= top_thresh, cur$gene, NA)
  cur$anno <- ifelse(cur$p_val_adj <= 0.05 & cur$avg_log2FC <= bottom_thresh, cur$gene, cur$anno)
  cur$color <- ifelse(cur$p_val_adj > 0.05, 'gray', ifelse(cur$avg_log2FC > 0, color1, color2))
  cur
}))

groups <- unique(degs$group)
groups <- group_levels[group_levels %in% groups]
plot_list <- list()
for(cluster  in groups){


  print(cluster)
  plot_degs <- cur_degs %>% subset(group == cluster)

  p <- plot_degs  %>%
     ggplot(aes(x=avg_log2FC, y=-log10(p_val_adj))) +
     geom_hline(yintercept=-log10(0.05), linetype='dashed')

  # plot genes that are Nr4a2 targets
  p <- p + ggrastr::rasterise(geom_point(
    alpha=0.5,
    color=plot_degs %>% .$color
  ), dpi=500)

  p <- p +
     geom_point(
       inherit.aes=FALSE,
       data=subset(plot_degs, !is.na(anno)),
       aes(avg_log2FC, -log10(p_val_adj)),
       fill=subset(plot_degs, !is.na(anno)) %>% .$color,
       shape=21, size=3, color='black'
     ) +
     geom_text_repel(aes(label=anno), color='black', fontface='italic',  min.segment.length=0, max.overlaps=Inf) +
     xlim(-1*max(abs(plot_degs$avg_log2FC))-0.1, max(abs(plot_degs$avg_log2FC))+0.1) +
     ggtitle(paste0(cluster)) +
     xlab(bquote("Average log"[2]~"(Fold Change)")) +
     ylab(bquote("-log"[10]~"(Adj. P-value)")) +
     theme(
       panel.border = element_rect(color='black', fill=NA, size=1),
       panel.grid.major = element_blank(),
       axis.line = element_blank(),
       plot.title = element_text(hjust = 0.5),
       legend.position='bottom'
     )

    plot_list[[cluster]] <- p

}


out <- paste0(fig_dir, 'volcano_', name, '_snRNA.pdf')


plot_list <- lapply(plot_list, function(x){
  x + theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    plot.margin = margin(0,0,0,0),
    plot.title = element_text(vjust=-0.2)
  )
})

pdf(out, width=15, height=18, useDingbats=FALSE)
wrap_plots(plot_list, ncol=5)
dev.off()

```

Run EnrichR for different sets of genes for each group

* Up in both
* Up in AD only
* Up in ADDS only
* Down in both
* Down in AD only
* Down in ADDS only

```{r eval=FALSE}

library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021', 'WikiPathway_2021_Human', 'KEGG_2021_Human')


# load DSAD vs Control DEGs:
degs_ad <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/AD_ci_degs.csv'))
degs_adds <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/DS_ci_FCX_degs.csv'))

# add gene_id column to both:
ix <- match(degs_ad$gene, kb_gene_table$gene_name)
degs_ad$gene_id <- kb_gene_table$gene_id[ix]

ix <- match(degs_adds$gene, sp_gene_table$gene_name)
degs_adds$gene_id <- sp_gene_table$gene_id[ix]

# genes in both:
genes.keep <- intersect(kb_gene_table$gene_id, sp_gene_table$gene_id)
degs_ad <- subset(degs_ad, gene_id %in% genes.keep)
degs_adds <- subset(degs_adds, gene_id %in% genes.keep)

# plot settings
groups <- unique(degs_ad$group)
logfc_thresh <- 0.05
combined_output <- data.frame()

# run loop over each group
for(cur_group in groups){
  print(cur_group)

  cur_degs_ad <- subset(degs_ad, group == cur_group)
  cur_degs_adds <- subset(degs_adds, group == cur_group)

  genes.keep <- intersect(cur_degs_ad$gene_id, cur_degs_adds$gene_id)
  cur_degs_ad <- subset(cur_degs_ad, gene_id %in% genes.keep)
  cur_degs_adds <- subset(cur_degs_adds, gene_id %in% genes.keep)

  # make sure they are in the same order:
  rownames(cur_degs_ad) <- cur_degs_ad$gene_id
  rownames(cur_degs_adds) <- cur_degs_adds$gene_id
  cur_degs_adds <- cur_degs_adds[cur_degs_ad$gene_id,]

  # get gene sets for enrichr:
  up_ad <- cur_degs_ad %>% subset(avg_log2FC >= logfc_thresh & p_val_adj < 0.05) %>% .$gene
  down_ad <- cur_degs_ad %>% subset(avg_log2FC <= -logfc_thresh & p_val_adj < 0.05) %>% .$gene
  up_adds <- cur_degs_adds %>% subset(avg_log2FC >= logfc_thresh & p_val_adj < 0.05) %>% .$gene
  down_adds <- cur_degs_adds %>% subset(avg_log2FC <= -logfc_thresh & p_val_adj < 0.05) %>% .$gene

  # list of inputs to enrichr
  input_list <- list(
    up_ad = up_ad[!(up_ad %in% up_adds)],
    down_ad = down_ad[!(down_ad %in% down_adds)],
    up_adds = up_adds[!(up_adds %in% up_ad)],
    down_adds = down_adds[!(down_adds %in% down_ad)],
    up_both = intersect(up_ad, up_adds),
    down_both = intersect(down_ad, down_adds),
    up_ad_down_adds = intersect(up_ad, down_adds),
    up_adds_down_ad = intersect(up_adds, down_ad)
  )

  # size of lists
  lapply(input_list, function(x){
    print(length(x))
  })


  # run enrichr and combine outputs
  enriched_df <- do.call(rbind, lapply(names(input_list), function(x){
    if(length(input_list[[x]]) > 0){
      cur_enrich <- enrichr(input_list[[x]], dbs)
    } else{return(data.frame())}
    cur_df <- do.call(rbind, lapply(dbs, function(cur_db){
      df <- cur_enrich[[cur_db]]
      if(nrow(df) > 1){df$degs <- x; df$group <- cur_group; df$db <- cur_db}
      else{df <- data.frame()}
      df
    }))
  }))

  combined_output <- rbind(combined_output, enriched_df)

}

# write the output to a tsv
write.table(combined_output, file=paste0(data_dir, 'snRNA_AD_ADDS_shared_distinct_GO_terms.tsv'), quote=FALSE, row.names=FALSE, sep='\t')

combined_output %>%
  subset(P.value < 0.05) %>%
  write.table(
    file=paste0(data_dir, 'snRNA_AD_ADDS_shared_distinct_GO_terms_signif.tsv'),
    quote=FALSE, row.names=FALSE, sep='\t'
  )


```


Plot selected GO terms:

```{r eval=FALSE}

color_df <- read.csv(file=paste0(data_dir, 'cell_identity.csv'))
color_df$group <- factor(
  as.character(color_df$group),
  levels = as.character(color_df$group)
)

# helper function to wrap text
wrapText <- function(x, len) {
    sapply(x, function(y) paste(strwrap(y, len), collapse = "\n"), USE.NAMES = FALSE)
}

combined_output <- read.delim('data/snRNA_AD_ADDS_shared_distinct_GO_terms.tsv', sep='\t', header=1)
selected_terms <- read.table('data/snRNA_shared_GO_upregulated_selected.txt', sep='\t', header=1)

# subset selected terms
selected_terms <- subset(combined_output, Term %in% selected_terms$Term & P.value < 0.005)

selected_terms$group <- factor(
  as.character(selected_terms$group),
  levels = color_df$group
)

# remove GO Term ID
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")

selected_terms <- selected_terms %>%
  arrange(group)


selected_terms$wrap <- wrapText(selected_terms$Term, 35)

selected_terms$Term <- factor(
  as.character(selected_terms$Term),
  levels = rev(unique(as.character(selected_terms$Term)))
)

# GO Term dot plot

p <- selected_terms %>%
  ggplot(aes(x = group, y = Term, color = -log(P.value), size=log(Odds.Ratio))) +
  geom_point() +
  scale_color_stepsn(colors=rev(magma(256))) +
  RotatedAxis() + xlab('') + ylab('') +
  theme(
    axis.title.x = element_blank(),
    axis.title.y = element_blank(),
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    axis.text.x = element_blank(),
    axis.ticks.x = element_blank(),
    plot.margin = margin(0,0,0,0),
    panel.grid = element_line(size=0.25, color='lightgrey')
  )


  # make the colorbar as its own heatmap
  color_df$var <- 1
  colorbar <- color_df %>%
    ggplot(aes(x=group, y=var, fill=group)) +
    geom_tile() +
    scale_fill_manual(values=color_df$colour) +
    coord_equal() +
    NoLegend() + RotatedAxis() +
    theme(
      plot.title=element_blank(),
      axis.line=element_blank(),
      axis.ticks.y =element_blank(),
      axis.text.y = element_blank(),
      axis.title = element_blank(),
      plot.margin=margin(0,0,0,0),
    )




pdf(paste0(fig_dir, 'selected_GO_terms.pdf'), width=13, height=13)
p / colorbar #+ plot_layout(heights=c(20,1))
dev.off()




```




Test Libra DEGs

```{r eval=FALSE}

library(Libra)

# set factor levels:
seurat_adds$Diagnosis <- factor(
  as.character(seurat_adds$Diagnosis),
  levels = c('DSAD', 'Control')
)

X <- GetAssayData(seurat_adds, slot='counts')
meta <- seurat_adds@meta.data

meta$cell_identity <- as.character(meta$cell_identity)

# PCC only:
cur_cells <- seurat_adds$Region == "PCC"
X <- X[,cur_cells]
meta <- meta[cur_cells,]

seurat_obj$Diagnosis <- factor(
  as.character(seurat_obj$Diagnosis),
  levels = c('ADDS', 'Control')
)

DE <- Libra::run_de(
  X,
  meta = meta,
  cell_type_col = "cell_identity",
  replicate_col = "Sample",
  label_col = "Diagnosis",
  de_family = 'pseudobulk',
  de_method = 'edgeR',
  de_type = 'LRT',
  min_features=0
)

pseudobulks = Libra::to_pseudobulk(
  input = X,
  meta = meta,
  replicate_col = replicate_col,
  cell_type_col = cell_type_col,
  label_col = label_col
)

# save:
write.csv(DE, quote=FALSE, file=paste0(data_dir, 'test_libra_DEGs1.csv'))
DE <- read.csv(file=paste0(data_dir, 'test_libra_DEGs1.csv'))

##############
# my own edgeR
##############
library(magrittr)
library(Libra)
library(edgeR)


edgeR_test <- pseudobulk_edgeR(
  seurat_adds,
  cell_type_col = 'cell_identity',
  label_col = 'Diagnosis',
  replicate_col = 'Sample',
  # covariates = c("Batch"),
  cells_use = seurat_adds$Region == 'FCX'
)

write.csv(edgeR_test, quote=FALSE, file=paste0(data_dir, 'test_edgeR_DEGs_FCX.csv'))
DE <- edgeR_test


################################################################################
# Make volcanos
################################################################################

nlabel <- 5
color1 <- 'darkgoldenrod3'; color2 <- 'hotpink3'

# label the top and bottom significant genes by log fold change
cur_degs <- Reduce(rbind, lapply(unique(DE$cell_type), function(x){
  cur <- subset(DE, cell_type == x)

  top_thresh <- cur %>% subset(p_val_adj <= 0.05) %>% top_n(nlabel, wt=avg_logFC) %>% .$avg_logFC %>% min
  bottom_thresh <- cur %>% subset(p_val_adj <= 0.05) %>% top_n(-1*nlabel, wt=avg_logFC) %>% .$avg_logFC %>% max

  cur$anno <- ifelse(cur$p_val_adj <= 0.05 & cur$avg_logFC >= top_thresh, cur$gene, NA)
  cur$anno <- ifelse(cur$p_val_adj <= 0.05 & cur$avg_logFC <= bottom_thresh, cur$gene, cur$anno)
  cur$color <- ifelse(cur$p_val_adj > 0.05, 'gray', ifelse(cur$avg_logFC > 0, color1, color2))
  cur
}))

groups <- unique(DE$cell_type)
plot_list <- list()
for(cluster  in groups){


  print(cluster)
  plot_degs <-  cur_degs %>% subset(cell_type == cluster)

  p <- plot_degs %>%
     ggplot(aes(x=avg_logFC, y=-log10(p_val_adj))) +
     geom_hline(yintercept=-log10(0.05), linetype='dashed')

  p <- p + ggrastr::rasterise(geom_point(alpha=0.5, color=plot_degs$color), dpi=500)

  p <- p +
     geom_point(inherit.aes=FALSE, data=subset(plot_degs, !is.na(anno)), aes(avg_logFC, -log10(p_val_adj)),fill=subset(plot_degs, !is.na(anno)) %>% .$color, shape=21, size=2, color='black') +
     geom_text_repel(aes(label=anno), color='black', fontface='italic',  min.segment.length=0) +
     xlim(-1*max(abs(plot_degs$avg_logFC))-0.1, max(abs(plot_degs$avg_logFC))+0.1) +
     ggtitle(paste0(cluster)) +
     theme(
       panel.grid.major = element_blank(),
       plot.title = element_text(hjust = 0.5),
       legend.position='bottom'
     )

    plot_list[[cluster]] <- p

}

pdf(paste0(fig_dir, 'edgeR_volcano_FCX.pdf'), width=24, height=20, useDingbats=FALSE)
wrap_plots(plot_list, ncol=6)
dev.off()


test <- subset(DE, cell_type == 'T-Pericyte' & p_val_adj < 0.05)

head(test)
table(abs(test$avg_logFC) >= 1)



subset(DE, gene == 'APP')

```

Correlation of MAST DEGs with edgeR DEGs


```{r eval=FALSE}

edgeR_DEGs <- DE
edgeR_DEGs <- as.data.frame(edgeR_DEGs) %>% dplyr::rename(c(avg_log2FC = avg_logFC, group = cell_type))

MAST_DEGs <- read.csv(paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/splitseq/integration/DEGs/combined/DS_ci_PCC_degs.csv'))




# plot settings
groups <- unique(edgeR_DEGs$group)
plot_list <- list()
plot_list_pval <- list()
rrho_plot_list <- list()
cor_list <- c()
signif_only <- FALSE
run_rrho <- FALSE
rrho_maxval <- 750

# run loop over each group
# initialize progress bar
pb <- utils::txtProgressBar(min = 0, max = length(groups), style = 3, width = 50, char = "=")
counter <- 1


for(cur_group in groups){
  print(cur_group)
  setTxtProgressBar(pb, counter)


  cur_MAST_DEGs <- subset(MAST_DEGs, group == cur_group)
  cur_edgeR_DEGs <- subset(edgeR_DEGs, group == cur_group)

  genes.keep <- intersect(cur_MAST_DEGs$gene, cur_edgeR_DEGs$gene)
  cur_MAST_DEGs <- subset(cur_MAST_DEGs, gene %in% genes.keep)
  cur_edgeR_DEGs <- subset(cur_edgeR_DEGs, gene %in% genes.keep)

  # make sure they are in the same order:
  rownames(cur_MAST_DEGs) <- cur_MAST_DEGs$gene
  rownames(cur_edgeR_DEGs) <- cur_edgeR_DEGs$gene
  cur_edgeR_DEGs <- cur_edgeR_DEGs[cur_MAST_DEGs$gene,]

  # join the two dataframes
  plot_df <- dplyr::inner_join(cur_MAST_DEGs, cur_edgeR_DEGs, by = 'gene')
  library(ggpubr)
  p1 <- ggpubr::ggscatter(
    plot_df,
    x = 'avg_log2FC.x',
    y = 'avg_log2FC.y',
    add = 'reg.line',
    add.params = list(color='blue', fill='lightgrey'),
  ) + stat_cor(method='pearson')


  logfc_thresh <- 0.05
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Consistent", "")
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Consistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x >= logfc_thresh & plot_df$avg_log2FC.y <= -logfc_thresh, "Inconsistent", plot_df$group)
  plot_df$group <- ifelse(plot_df$avg_log2FC.x <= -logfc_thresh & plot_df$avg_log2FC.y >= logfc_thresh, "Inconsistent", plot_df$group)
  group_colors <- c('grey', 'blue', 'yellow')

  if(signif_only){
    plot_df <- subset(plot_df, p_val_adj.x < 0.05 | p_val_adj.y < 0.05)
    if(nrow(plot_df) == 0){next}
  }

  cur_cor <-  cor(x=as.numeric(plot_df$avg_log2FC.x), y=as.numeric(plot_df$avg_log2FC.y))
  cor_list <- c(cor_list, cur_cor)
  print(table(plot_df$group))
  print(dim(plot_df))

  plot_range <- max(max(plot_df$avg_log2FC.x), max(plot_df$avg_log2FC.y))
  p <- plot_df %>%
    ggplot(aes(x = avg_log2FC.x, y = avg_log2FC.y, color=group)) +
    geom_hline(yintercept = 0, linetype='dashed', color='grey') +
    geom_vline(xintercept = 0, linetype='dashed', color='grey') +
    ggrastr::rasterise(geom_point(), dpi=500) +
    scale_color_manual(values=group_colors) +
    geom_smooth(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='lm', color='black') +
    stat_cor(inherit.aes=FALSE, data=plot_df, mapping = aes(x = avg_log2FC.x, y = avg_log2FC.y), method='pearson') +
    xlim(c(-plot_range, plot_range)) +
    ylim(c(-plot_range, plot_range)) +
    theme(
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      plot.title = element_text(hjust=0.5)
    ) +
    coord_fixed(ratio=1) + NoLegend() +
    xlab(bquote("MAST Average log"[2]~"(Fold Change)")) +
    ylab(bquote("edgeR Average log"[2]~"(Fold Change)")) +
    ggtitle(cur_group)

  plot_list[[cur_group]] <- p

  plot_range <- max(
    max(-log(plot_df$p_val_adj.x[plot_df$p_val_adj.x != 0])),
     max(-log(plot_df$p_val_adj.y[plot_df$p_val_adj.y != 0]))
   )

  p <- plot_df %>%
    ggplot(aes(x = -log(p_val_adj.x), y = -log(p_val_adj.y), color=group)) +
    geom_hline(yintercept = 0, linetype='dashed', color='grey') +
    geom_vline(xintercept = 0, linetype='dashed', color='grey') +
    ggrastr::rasterise(geom_point(), dpi=500) +
    scale_color_manual(values=group_colors) +
    geom_smooth(inherit.aes=FALSE, data=plot_df, mapping = aes(x = -log(p_val_adj.x), y =-log(p_val_adj.y)), method='lm', color='black') +
    stat_cor(inherit.aes=FALSE, data=plot_df, mapping = aes(x = -log(p_val_adj.x), y = -log(p_val_adj.y)), method='pearson') +
    xlim(c(0, plot_range)) +
    ylim(c(0, plot_range)) +
    theme(
      axis.line.x = element_blank(),
      axis.line.y = element_blank(),
      panel.border = element_rect(colour = "black", fill=NA, size=1),
      plot.title = element_text(hjust=0.5)
    ) +
    coord_fixed(ratio=1) + NoLegend() +
    xlab(bquote("MAST -log"[10]~"(Adj. P-value)")) +
    ylab(bquote("edgeR log"[10]~"(Adj. P-alue)")) +
    ggtitle(cur_group)

  plot_list_pval[[cur_group]] <- p

  # update progress bar
  counter <- counter+1
}

# close progress bar
close(pb)


names(cor_list) <- names(plot_list)
plot_list <- plot_list[rev(order(cor_list))]
plot_list_pval <- plot_list_pval[rev(order(cor_list))]

pdf(paste0(fig_dir, 'edgeR_vs_MAST_combined.pdf'), width=32, height=16)
wrap_plots(plot_list, ncol=8)
dev.off()

pdf(paste0(fig_dir, 'edgeR_vs_MAST_pvals_combined.pdf'), width=32, height=16)
wrap_plots(plot_list_pval, ncol=8)
dev.off()

pdf(paste0(fig_dir, 'edgeR_vs_MAST_combined.pdf'), width=7, height=7)
p
dev.off()


pdf(paste0(fig_dir, 'deg_rrho_combined.pdf'), width=12, height=10)
wrap_plots(rrho_plot_list, ncol=6) + plot_layout(guides='collect')
dev.off()


```
