
In this notebook, we will compute gene signature scores for known relevant  gene
signatures (DAMs, DAAs, etc), and we will also use the Voyager R package to try to
perform geospatial statistical analysis based on these features.

Load the Seurat data

```{r eval=FALSE}

# conda activate voyager
library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(RColorBrewer)
library(Voyager)
library(UCell)
library(ggrastr)
library(ggrepel)
library(viridis)
library(Voyager)
library(SpatialFeatureExperiment)
library(scater)
library(scran)
library(SFEData)
library(sf)
library(ggplot2)
library(scales)
library(patchwork)
library(BiocParallel)
library(bluster)
library(tidyverse)
library(ggpubr)
library(hdWGCNA)
theme_set(theme_cowplot())

# load helper functions
source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')

# setup folders
setwd("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/voyager/")
fig_dir <- "figures/"
data_dir <- "data/"

# re-load human SFE dataset and geospatial seurat
sfe <- readRDS('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/voyager/data/ADDS_SFE.rds')
seurat_human <- readRDS("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/voyager/data/ADDS_seurat_processed_geospatial.rds")

################################################################################
# Load data for this notebook
################################################################################

seurat_human <- readRDS("/dfs7/swaruplab/emiyoshi/Visium_ADDS/ADDS_seurat_processed_annotated.rds")

seurat_human$Diagnosis <- factor(
  as.character(seurat_human$Diagnosis),
  levels = c("Control", "earlyAD", "AD", "AD_DS")
)

# 5x data:
seurat_5x <- readRDS('/dfs7/swaruplab/emiyoshi/Visium_5X/5XFAD_seurat_processed_annotated.rds')

seurat_5x$Age <- factor(
  as.character(seurat_5x$Age),
  levels = c("4mo", "6mo", "8mo", "12mo")
)

human_cp <- c(
      "L1" = "#8B3D5A", "L2-3" = "#E7BDE1", "L3-4" = "#E6A4CD",
      "L3-4-5" = "#CF8BA3", "L5-6" = "#9E6D7F", "L6b" = "#CDAEB9", "WM1" = "#64BCDB", "WM2" = "#62A7D7", "WM3" = "#99C8D7")

mouse_cp <- c(
      "ctx-deep-layers" = "#5b4468", "ctx-upper-layers" = "#9581a4", "ctx-olfactory" = "#8073ab",
      "hippocampus" = "#d6a5b2", "hippocampus-pyramidal" = "#e193aa",
      "lateral-ventricle" = "#d2c3b2",
      "striatum" = "#F0DE8C",
      "thalamus1" = "#00CBA7", "thalamus2" = "#abcc94",
      "hypothalamus-amygdala" = "#72c9b1",
      "WM1" = "#325ea8", "WM2" = "#64bcdb", "WM-cerebral-peduncle" = "#62a7d7",
      "erythrocytes-neurons" = "#d07fc4", "unknown" = "#c47c4d"
    )


# set factor levels for human clusters:
seurat_human$annotation <- factor(
  as.character(seurat_human$annotation),
  levels = names(human_cp)
)

# set factor levels for mouse clusters
seurat_5x$annotation <- factor(
  as.character(seurat_5x$annotation),
  levels = names(mouse_cp)
)




#
# seurat objets with amyloid data:

seurat_amyloid_human <- readRDS('/dfs7/swaruplab/emiyoshi/Visium_ADDS/ADDS_seurat_processed_amyloid.rds')
seurat_amyloid_5x <- readRDS('/dfs7/swaruplab/emiyoshi/Visium_5X/5XFAD_seurat_processed_amyloid.rds')


```

Compute gene signatures in the human dataset
based on code that Emily sent me, and then plot the scores

```{r eval=FALSE}


# Chen 2021 (compilation of multiple studies) - human orthologs of mouse genes
homeostatic <- c("BIN1", "SIGLEC6", "CD33", "CSF1R",
              "CST3", "CX3CR1", "HEXB", "JUN",
              "MEF2A", "MS4A6A", "P2RY12", # "MS4A6E" error
              "P2RY13", "SALL1", "SELPLG", "SERINC3",
              "TGFBR1", "TMEM119")

DAM <- c("APOE", "AXL", "B2M", # "CCL3L3",
      "CCL3L1", "CCL3", "CCL18", "CCL23",
      "CD63", # "CCL15-CCL14", "CCL15",
      "CD9", "CLEC7A", "CSF1", "CST7",
      "CTSB", "CTSD", "FTH1", "GPNMB",
      "IGF1", "IRF8", "ITGAX", "LGALS3",
      "LPL", "SPP1", "TIMP2", "TREM2",
      "TYROBP", "CD68")

IFN <- c("CD69", "CXCL10", "IFIT2", "IFIT3",
      "IRF7", "ISG15", "USP41", "USP18",
      "IFITM3", "STAT1")

MHC <- c("CD74", "HLA-DQB2", "HLA-DQB1", "HLA-E",
      "HLA-DMA", "HLA-A", "HLA-DRB1", "HLA-DRA",
      "HLA-DPB1", "CD83", "CD81")

Cyc_M <- c("BIRC5", "CENPE", "MCM5", "MKI67", "TOP2A")

GFAP_low <- c("LUZP2", "SLC7A10", "MFGE8")
GFAP_high <- c("GFAP", "ID3", "AQP4", "MYOC", "ID1", "FABP7")
DAA <- c("GFAP", "CTSB", "VIM", "OSMR", "SERPINA3", "GSN")


# Kenigsbuch et al 2022 - human orthologs of mouse genes
DOL <- c("B2M", "C4B", "C4A", "CD63",
      "CD9", "CLPTM1", "CTSB", "FABP5",
      "GPD1", "GSTP1", "HLA-E", # "H3-3A",
      "IL33", "KLK6", "LBH", "OPALIN",
      "PLEKHA1", "RNASE4", "RPL26", "RPS2",
      "SERPINA3", "SGK1", "STMN1")

DOL_markers <- list("DOL" = DOL, 'NFOL'=c('TCF7L2', 'CASR', 'CEMIP2', 'ITPR2'),
								  'MFOL'=c('MAL', 'MOG', 'PLP1', 'OPALIN', 'SERINC5', 'CTPS1'),
								  'MOL'=c('KLK6', 'APOD', 'SLC5A11', 'PDE1A'))

# load mouse <-> human gene name table:
hg38_mm10_genes <- read.table(
  "/dfs7/swaruplab/smorabit/resources/hg38_mm10_orthologs_2021.txt",
  sep='\t',
  header=TRUE
)
colnames(hg38_mm10_genes) <-c('hg38_id', 'mm10_id', 'mm10_name', 'hg38_name')
hg38_mm10_genes <- dplyr::select(hg38_mm10_genes, c(hg38_name, mm10_name, hg38_id, mm10_id))

hg38_mm10_genes <- subset(hg38_mm10_genes, mm10_name != '' & hg38_name != '')

# need to make sure that there's only one entry for each gene in hg38_mm10_genes
mm10_genes <- unique(hg38_mm10_genes$mm10_name)
hg38_genes <- unique(hg38_mm10_genes$hg38_name)
hg38_mm10_genes <- hg38_mm10_genes[match(mm10_genes, hg38_mm10_genes$mm10_name),]

# Load PIGs from Chen et al 2020
pig_table <- read.csv('~/swaruplab/smorabit/analysis/ADDS_2021/data/chen_modules.csv')

# convert to human names
ix <- match(pig_table$gene, hg38_mm10_genes$mm10_name)
pig_table$gene_hg38 <- hg38_mm10_genes$hg38_name[ix]

pigs <- subset(pig_table, module == 'PIG') %>% .$gene_hg38 %>% na.omit %>% as.character
oligs <- subset(pig_table, module == 'OLIG') %>% .$gene_hg38 %>% na.omit %>% as.character


gene_lists <- list("homeostatic" = homeostatic, "DAM" = DAM, "IFN" = IFN, "MHC" = MHC, "Cyc_M" = Cyc_M, 'PIGs' = pigs, 'OLIG' = oligs)
gene_lists <- c(gene_lists, DOL_markers)

# compute scores:
seurat_human <- AddModuleScore_UCell(seurat_human, features=gene_lists)


feats <- paste0(names(gene_lists), '_UCell')

for(cur_gene in feats){
  p <- SampleFeaturePlot(
    seurat_human,
    feature=cur_gene,
    #samples_to_plot = selected_samples,
    sample_labels = c("Diagnosis", "Sex", 'Age'),
    ncol = 10,
    raster=TRUE,
    plot_max = 'q99',
    plot_min = 'q5',
    colfunc = inferno,
    rev_colors=TRUE,
    dpi=800,
    combine=FALSE
  )

  p <- lapply(p, function(x){
    x + theme(
      plot.title = element_text(face='bold', size=15, vjust=-1),
      plot.margin = margin(0,0,0,0)
    ) + labs(fill = paste0(feat_name, ' ', modifier))
  })

  patch <- wrap_plots(p, ncol=8, widths=1, heights=1) + plot_layout(guides='collect')

  pdf(paste0(fig_dir, 'genescores/', cur_gene, '_featureplot.pdf'), width=12, height=12)
#  print(p)
  print(patch)
  dev.off()

}


```

Test setting up a SpatialFeatureExperiment for one sample:

```{r eval=FALSE}

setwd('voyager/')



kb_gene_table <- read.table("/dfs7/swaruplab/shared_lab/cross-disorder/count_matrices/AD_Mathys_2019/D17-8777/counts_unfiltered/genes.tsv")

gene_file <- '~/swaruplab/smorabit/data/ADDS_2021/visium/Dec_13_2021/spaceranger_count/Human8/outs/filtered_feature_bc_matrix/features.tsv.gz'
gene_table <- read.table(gene_file, sep='\t')
all.equal(gene_table$V1, rownames(sfe))

# representative samples
human_rep_samples <- c(
  'Dec_13_2021_Human5', 'Dec_20_2021_Human1',
  'Dec_13_2021_Human6', 'Oct_2021_6',
  'Dec_13_2021_Human3', 'Dec_13_2021_Human7',
  'Nov_24_2021_VisiumHuman_12', 'Dec_13_2021_Human8'
)

cur_vis_sample <- 'Dec_13_2021_Human8'

# get the current visium sample and make sure to set up the image
cur_vis <- seurat_human[,seurat_human@meta.data[["Sample"]] == cur_vis_sample]
cur_image <- names(cur_vis@images)[sapply(names(cur_vis@images), function(x){nrow(cur_vis@images[[x]]@coordinates) > 0})]
cur_vis@images <- list(cur_image = cur_vis@images[[cur_image]])

# get the barcode:
cur_vis$bc <- do.call(rbind, strsplit(colnames(cur_vis), '_'))[,1]

######################################################################
# make a sfe object from the spaceranger output:
#
# note: I tried to make it from scratch and it didn't work
######################################################################

file <- '~/swaruplab/smorabit/data/ADDS_2021/visium/Dec_13_2021/spaceranger_count/Human8/outs/'
sfe <- read10xVisiumSFE(file, type='sparse', data='raw', load=FALSE)

rownames(sfe) <- gene_table$V2

# only keep spots that are in the tissue
sfe <- sfe[, colData(sfe)$in_tissue]
#sfe <- sfe[rowSums(counts(sfe)) > 0,]

all.equal(as.character(colnames(sfe)), as.character(cur_vis$bc))

meta <- cur_vis@meta.data
rownames(meta) <- colnames(sfe)
colData(sfe) <- cbind(colData(sfe), meta)

clusters <- sfe$annotation
#clusters <- quickCluster(sfe)
sfe <- computeSumFactors(sfe, clusters=clusters)
#> Warning in (function (x, sizes, min.mean = NULL, positive = FALSE, scaling =
#> NULL) : encountered non-positive size factor estimates
sfe <- sfe[, sizeFactors(sfe) > 0]
sfe <- logNormCounts(sfe)
dim(sfe)



# spatial neighborhood graph
colGraph(sfe, "visium") <- findVisiumGraph(sfe)

sfe$myfeat <- sfe$DAM_UCell


p <-  plotColGraph(sfe, colGraphName = "visium", colGeometryName = "spotPoly")
pdf(paste0(fig_dir, 'spat_graph.pdf'), width=5, height=5)
p
dev.off()


################################################################################
# Compute Local Moran's I & Getis-Ord(*) for different features
################################################################################

# compute a binary-weight graph
colGraph(sfe, "visium_B") <- findVisiumGraph(sfe, style = "B")

# compute log scale of OC and Amyloglo scores
colData(sfe)$Amyloglo_Score_log <- log(sfe$Amyloglo_Score + 1)
colData(sfe)$OC_Score_log <- log(sfe$OC_Score + 1)

# select features
feats <- c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell', 'Amyloglo_Score_log', 'OC_Score_log')

# compute Moran's I
sfe <- colDataUnivariate(sfe, type = "localmoran", features = feats,
                            colGraphName = "visium")

# compute Getis-Ord Gi*
sfe <- colDataUnivariate(sfe, type = "localG_perm", features = feats,
                            colGraphName = "visium_B", include_self=TRUE)


for(cur_feat in feats){
  print(cur_feat)
  p1 <- plotSpatialFeature(sfe, features = cur_feat, colGeometryName = "spotPoly") +
    RotatedAxis() + ggtitle(cur_feat) + labs(fill='') + umap_theme()
  p2 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
                  colGeometryName = "spotPoly",divergent = TRUE,
                  diverge_center = 0) +
                    RotatedAxis() + labs(fill='') + umap_theme()
  p3 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
                  colGeometryName = "spotPoly", attribute = "-log10p_adj", divergent = TRUE,
                  diverge_center = -log10(0.05)) +
                    RotatedAxis() + labs(fill='') + umap_theme()

  p4 <- plotLocalResult(sfe, "localG_perm", features = cur_feat,
                  colGeometryName = "spotPoly",divergent = TRUE,
                  diverge_center = 0) +
                  ggtitle('Getis-Ord Gi(*)') +
                  RotatedAxis() + labs(fill='') + umap_theme()

  # p4 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
  #                 colGeometryName = "spotPoly", attribute = "pysal") +
  #                 RotatedAxis() + labs(fill='') + umap_theme()

  pdf(paste0(fig_dir, 'plot_', cur_feat,'.pdf'), width=8, height=6)
  print((p1 + p2) / (p3 + p4))
  dev.off()
}


```

Try spatial analysis of Amyloglo and OC scores with Voyager in one sample

```{r eval=FALSE}

table(seurat_amyloid_human$Sample)

cur_vis_sample <- 'Dec_13_2021_Human8'

# get the current visium sample and make sure to set up the image
cur_vis <- seurat_amyloid_human[,seurat_amyloid_human@meta.data[["Sample"]] == cur_vis_sample]
cur_image <- names(cur_vis@images)[sapply(names(cur_vis@images), function(x){nrow(cur_vis@images[[x]]@coordinates) > 0})]
cur_vis@images <- list(cur_image = cur_vis@images[[cur_image]])

# get the barcode:
cur_vis$bc <- do.call(rbind, strsplit(colnames(cur_vis), '_'))[,1]
all.equal(as.character(colnames(sfe)), as.character(cur_vis$bc))

# add scores:
meta <- cur_vis@meta.data[,c('Amyloglo_Score', 'OC_Score', 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell')]
rownames(meta) <- colnames(sfe)
colData(sfe) <- cbind(colData(sfe), meta)


################################################################################
# Correlate moran's I for OC and Amylo-glo scores:
################################################################################

feat1 <- 'Amyloglo_Score_log'
feat2 <- 'OC_Score_log'


meta <- as.data.frame(colData(sfe))
meta$feat1 <- as.numeric(localResult(sfe, 'localmoran', feat1)$Ii)
meta$feat2 <- as.numeric(localResult(sfe, 'localmoran', feat2)$Ii)

test <-  localResult(sfe, 'localmoran', 'Amyloglo_Score_log')

p <- meta %>%
  ggplot(aes(x = feat1, y = feat2, color = annotation)) +
  geom_hline(yintercept=0, linetype='dashed', color='grey90') +
  geom_vline(xintercept=0, linetype='dashed', color='grey90') +
  geom_point(alpha=0.2) +
  geom_smooth(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2, color=annotation), method='lm') +
  stat_cor(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2), method='pearson') +
  #xlim(c(-plot_range, plot_range)) +
  #ylim(c(-plot_range, plot_range)) +
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, linewidth=1),
    plot.title = element_text(hjust=0.5)
  ) + NoLegend() +
  scale_color_manual(values=human_cp)

p <- p +
xlab("Amyloglo local Moran's I") +
ylab("OC local Moran's I")

pdf(paste0(fig_dir, 'test_OC_amylo_corr.pdf'), width=6, height=6)
p + facet_wrap(~ annotation, ncol=3)
dev.off()



################################################################################
# Correlate moran's I for OC and DAM scores
################################################################################

feat1 <- 'DAM_UCell'
feat2 <- 'OC_Score_log'


meta <- as.data.frame(colData(sfe))
meta$feat1 <- as.numeric(localResult(sfe, 'localmoran', feat1)$Ii)
meta$feat2 <- as.numeric(localResult(sfe, 'localmoran', feat2)$Ii)


p <- meta %>%
  ggplot(aes(x = feat1, y = feat2, color = annotation)) +
  geom_hline(yintercept=0, linetype='dashed', color='grey90') +
  geom_vline(xintercept=0, linetype='dashed', color='grey90') +
  geom_point(alpha=0.2) +
  geom_smooth(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2, color=annotation), method='lm') +
  stat_cor(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2), method='pearson') +
  #xlim(c(-plot_range, plot_range)) +
  #ylim(c(-plot_range, plot_range)) +
  theme(
    axis.line.x = element_blank(),
    axis.line.y = element_blank(),
    panel.border = element_rect(colour = "black", fill=NA, linewidth=1),
    plot.title = element_text(hjust=0.5)
  ) + NoLegend() +
  scale_color_manual(values=human_cp)

p <- p +
xlab("DAM local Moran's I") +
ylab("OC local Moran's I")

pdf(paste0(fig_dir, 'test_OC_DAM_corr.pdf'), width=6, height=6)
p + facet_wrap(~ annotation, ncol=3)
dev.off()




################################################################################
# Correlate Getis-Ord Gi(*) for DAM & OC
################################################################################

feat1 <- 'DAM_UCell'
feat2 <- 'OC_Score_log'

feats <- c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell', 'Amyloglo_Score_log', 'OC_Score_log')

for(amyloid in c('Amyloglo_Score_log', 'OC_Score_log')){
  for(feat in c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell')){

    print(paste0(feat, ' ', amyloid))
    meta <- as.data.frame(colData(sfe))
    meta$feat1 <- as.numeric( localResult(sfe, 'localG_perm', amyloid)[,'localG'])
    meta$feat2 <- as.numeric( localResult(sfe, 'localG_perm', feat)[,'localG'])

    p <- meta %>%
      ggplot(aes(x = feat1, y = feat2, color = annotation)) +
      geom_hline(yintercept=0, linetype='dashed', color='grey90') +
      geom_vline(xintercept=0, linetype='dashed', color='grey90') +
      geom_point(alpha=0.2) +
      geom_smooth(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2, color=annotation), method='lm') +
      stat_cor(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2), method='pearson') +
      #xlim(c(-plot_range, plot_range)) +
      #ylim(c(-plot_range, plot_range)) +
      theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, linewidth=1),
        plot.title = element_text(hjust=0.5)
      ) + NoLegend() +
      scale_color_manual(values=human_cp)

    p <- p +
    xlab(paste0(amyloid, " Getis-Ord Gi(*)")) +
    ylab(paste0(feat, " Getis-Ord Gi(*)"))

    pdf(paste0(fig_dir, 'corr_', amyloid, '_', feat, '_Gi_corr.pdf'), width=6, height=6)
    print(p + facet_wrap(~ annotation, ncol=3))
    dev.off()

  }
}



```


Test setting up a SpatialFeatureExperiment for one mouse sample

```{r eval=FALSE}

setwd('voyager/')

library(Voyager)
library(SpatialFeatureExperiment)
library(scater)
library(scran)
library(SFEData)
library(sf)
library(ggplot2)
library(scales)
library(patchwork)
library(BiocParallel)
library(bluster)
library(tidyverse)
library(ggpubr)


dir.create('figures/5x')
fig_dir <- 'figures/5x/'

# which sample to pick?
names(seurat_amyloid_5x@meta.data)
table(seurat_amyloid_5x$SAMPLE, seurat_amyloid_5x$Age)
table(seurat_amyloid_5x$SampleID)


gene_file <- '~/swaruplab/smorabit/data/Visium_Mouse_2021/Dec_20_2021/spaceranger_count/5XFAD28/outs/filtered_feature_bc_matrix/features.tsv.gz'
gene_table <- read.table(gene_file, sep='\t')


cur_vis_sample <- '5101'

# get the current visium sample and make sure to set up the image
cur_vis <- seurat_amyloid_5x[,seurat_amyloid_5x@meta.data[["SAMPLE"]] == cur_vis_sample]
cur_image <- names(cur_vis@images)[sapply(names(cur_vis@images), function(x){nrow(cur_vis@images[[x]]@coordinates) > 0})]
cur_vis@images <- list(cur_image = cur_vis@images[[cur_image]])

# get the barcode:
cur_vis$bc <- do.call(rbind, strsplit(colnames(cur_vis), '_'))[,1]

######################################################################
# make a sfe object from the spaceranger output:
#
# note: I tried to make it from scratch and it didn't work
######################################################################

file <- '~/swaruplab/smorabit/data/Visium_Mouse_2021/Dec_20_2021/spaceranger_count/5XFAD28/outs'
sfe <- read10xVisiumSFE(file, type='sparse', data='raw', load=FALSE)

# change to gene names
all.equal(gene_table$V1, rownames(sfe))
rownames(sfe) <- gene_table$V2

# only keep spots that are in the tissue
sfe <- sfe[, colData(sfe)$in_tissue]
#sfe <- sfe[rowSums(counts(sfe)) > 0,]

# same size as the Seurat object?
dim(sfe)
dim(cur_vis)
all.equal(as.character(colnames(sfe)), as.character(cur_vis$bc))

meta <- cur_vis@meta.data
rownames(meta) <- colnames(sfe)
colData(sfe) <- cbind(colData(sfe), meta)

# log normalize data
clusters <- sfe$annotation
sfe <- computeSumFactors(sfe, clusters=clusters)
sfe <- logNormCounts(sfe)

# compute spatial neighborhood graph
colGraph(sfe, "visium") <- findVisiumGraph(sfe)

# compute a binary-weight graph
colGraph(sfe, "visium_B") <- findVisiumGraph(sfe, style = "B")


################################################################################
# Compute Local Moran's I & Getis-Ord(*) for different features
################################################################################

# compute log scale of OC and Amyloglo scores
colData(sfe)$Amyloglo_Score_log <- log(sfe$Amyloglo_Score + 1)
colData(sfe)$OC_Score_log <- log(sfe$OC_Score + 1)

# select features
feats <- c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell', 'Amyloglo_Score_log', 'OC_Score_log')

# compute Moran's I
sfe <- colDataUnivariate(sfe, type = "localmoran", features = feats,
                            colGraphName = "visium")

# compute Getis-Ord Gi*
sfe <- colDataUnivariate(sfe, type = "localG_perm", features = feats,
                            colGraphName = "visium_B", include_self=TRUE)


for(cur_feat in feats){
  print(cur_feat)
  p1 <- plotSpatialFeature(sfe, features = cur_feat, colGeometryName = "spotPoly") +
    RotatedAxis() + ggtitle(cur_feat) + labs(fill='') + umap_theme()
  p2 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
                  colGeometryName = "spotPoly",divergent = TRUE,
                  diverge_center = 0) +
                    RotatedAxis() + labs(fill='') + umap_theme()
  p3 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
                  colGeometryName = "spotPoly", attribute = "-log10p_adj", divergent = TRUE,
                  diverge_center = -log10(0.05)) +
                    RotatedAxis() + labs(fill='') + umap_theme()

  p4 <- plotLocalResult(sfe, "localG_perm", features = cur_feat,
                  colGeometryName = "spotPoly",divergent = TRUE,
                  diverge_center = 0) +
                  ggtitle('Getis-Ord Gi(*)') +
                  RotatedAxis() + labs(fill='') + umap_theme()

  # p4 <- plotLocalResult(sfe, "localmoran", features = cur_feat,
  #                 colGeometryName = "spotPoly", attribute = "pysal") +
  #                 RotatedAxis() + labs(fill='') + umap_theme()

  pdf(paste0(fig_dir, 'plot_', cur_feat,'_5x.pdf'), width=8, height=6)
  print((p1 + p2) / (p3 + p4))
  dev.off()
}


################################################################################
# Correlate Getis-Ord Gi(*) for DAM & OC
################################################################################


for(amyloid in c('Amyloglo_Score_log', 'OC_Score_log')){
  for(feat in c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell')){

    print(paste0(feat, ' ', amyloid))
    meta <- as.data.frame(colData(sfe))
    meta$feat1 <- as.numeric( localResult(sfe, 'localG_perm', amyloid)[,'localG'])
    meta$feat2 <- as.numeric( localResult(sfe, 'localG_perm', feat)[,'localG'])

    meta <- subset(meta, annotation!= 'unknown')

    p <- meta %>%
      ggplot(aes(x = feat1, y = feat2, color = annotation)) +
      geom_hline(yintercept=0, linetype='dashed', color='grey90') +
      geom_vline(xintercept=0, linetype='dashed', color='grey90') +
      geom_point(alpha=0.2) +
      geom_smooth(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2, color=annotation), method='lm') +
      stat_cor(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2), method='pearson') +
      #xlim(c(-plot_range, plot_range)) +
      #ylim(c(-plot_range, plot_range)) +
      theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, linewidth=1),
        plot.title = element_text(hjust=0.5)
      ) + NoLegend() +
      scale_color_manual(values=mouse_cp)

    p <- p +
    xlab(paste0(amyloid, " Getis-Ord Gi(*)")) +
    ylab(paste0(feat, " Getis-Ord Gi(*)"))

    pdf(paste0(fig_dir, 'GI_corr/corr_', amyloid, '_', feat, '_Gi_corr.pdf'), width=8, height=8)
    print(p + facet_wrap(~ annotation, ncol=4))
    dev.off()

  }
}


################################################################################
# Correlate Moran's I for DAM & OC
################################################################################


for(amyloid in c('Amyloglo_Score_log', 'OC_Score_log')){
  for(feat in c(paste0(names(gene_lists), '_UCell'), 'GFAP_high_UCell', 'DAA_UCell', 'GFAP_low_UCell')){

    print(paste0(feat, ' ', amyloid))
    meta <- as.data.frame(colData(sfe))
    meta$feat1 <- as.numeric( localResult(sfe, 'localmoran', amyloid)[,'Ii'])
    meta$feat2 <- as.numeric( localResult(sfe, 'localmoran', feat)[,'Ii'])

    meta <- subset(meta, annotation!= 'unknown')

    p <- meta %>%
      ggplot(aes(x = feat1, y = feat2, color = annotation)) +
      geom_hline(yintercept=0, linetype='dashed', color='grey90') +
      geom_vline(xintercept=0, linetype='dashed', color='grey90') +
      geom_point(alpha=0.2) +
      geom_smooth(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2, color=annotation), method='lm') +
      stat_cor(inherit.aes=FALSE, data=meta, mapping = aes(x = feat1, y = feat2), method='pearson') +
      #xlim(c(-plot_range, plot_range)) +
      #ylim(c(-plot_range, plot_range)) +
      theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(colour = "black", fill=NA, linewidth=1),
        plot.title = element_text(hjust=0.5)
      ) + NoLegend() +
      scale_color_manual(values=mouse_cp)

    p <- p +
    xlab(paste0(amyloid, " local Moran's I")) +
    ylab(paste0(feat, " local Moran's I"))

    pdf(paste0(fig_dir, 'LMI_corr/corr_', amyloid, '_', feat, '_LMI_corr.pdf'), width=8, height=8)
    print(p + facet_wrap(~ annotation, ncol=4))
    dev.off()

  }
}

```




Create an SFE object holding all 39 samples

```{r eval=FALSE}

setwd('voyager/')

library(Voyager)
library(SpatialFeatureExperiment)
library(scater)
library(scran)
library(SFEData)
library(sf)
library(ggplot2)
library(scales)
library(patchwork)
library(BiocParallel)
library(bluster)
library(tidyverse)
library(ggpubr)



table(seurat_human$Sample)
table(seurat_human$SampleID)
table(seurat_human$seqbatch)




# load visium data
spaceranger_dir <- '/dfs7/swaruplab/smorabit/data/ADDS_2021/visium/'
batches <- c('October_2021', 'Nov_24_2021', 'Dec_13_2021', 'Dec_20_2021')

files <- unlist(lapply(batches, function(x){
  cur_dir <- paste0(spaceranger_dir, x, '/spaceranger_count/')
  paste0(cur_dir, dir(cur_dir), '/outs/')
}))
files

samples <- unlist(lapply(batches, function(x){
  cur_dir <- paste0(spaceranger_dir, x, '/spaceranger_count/')
  paste0(x,'_', dir(cur_dir))
}))
samples <- gsub('October', 'Oct', samples)

# create SFE from visium data
sfe <- read10xVisiumSFE(files, sample_id=samples, type='sparse', data='raw', load=FALSE)

# only keep spots that are in the tissue
sfe <- sfe[, colData(sfe)$in_tissue]

##############################################################
# transfer the Seurat meta-data to the SFE object
##############################################################

# match barcodes between the seurat object and the SFE object
bc <- do.call(rbind, strsplit(colnames(seurat_human), '_'))[,1]
bc <- do.call(rbind, strsplit(bc, '-'))[,1]
seurat_human$bc <- paste0(bc, '-', as.character(seurat_human$Sample))
bc <- do.call(rbind, strsplit(colnames(sfe), '-'))[,1]
sfe$bc <- paste0(bc, '-', as.character(sfe$sample_id))

# match seurat metadata
meta <- seurat_human@meta.data
ix <- match(seurat_human$bc, sfe$bc)
meta <- meta[ix,] # %>% dplyr::select(-bc)
all.equal(as.character(sfe$bc), as.character(meta$bc))

meta <- meta[,!(colnames(meta) %in% colnames(colData(sfe)))]

# add Seurat metadata to SFE
colData(sfe) <- cbind(colData(sfe), meta)

##############################################################
# change gene names from ensembl ID to gene symbols
##############################################################

gene_file <- '~/swaruplab/smorabit/data/ADDS_2021/visium/Dec_13_2021/spaceranger_count/Human8/outs/filtered_feature_bc_matrix/features.tsv.gz'
gene_table <- read.table(gene_file, sep='\t')

# set gene names
rownames(sfe) <- gene_table$V2

##############################################################
# normalize data and compute spatial graphs
##############################################################

clusters <- sfe$annotation
sfe <- computeSumFactors(sfe, clusters=clusters)
sfe <- logNormCounts(sfe)

# compute spatial neighborhood graph
gs <- findVisiumGraph(sfe, sample_id = 'all')
for(sample in names(gs)){
  colGraph(sfe, sample_id = sample, 'visium') <- gs[[sample]]
}

# compute a binary-weight graph
gs <- findVisiumGraph(sfe, sample_id= 'all', style = "B")
for(sample in names(gs)){
  colGraph(sfe, sample_id = sample, 'visium_B') <- gs[[sample]]
}

##############################################################
# transfer the amyloid data to the sfe object
##############################################################

# match barcodes between the seurat object and the SFE object
bc <- do.call(rbind, strsplit(colnames(seurat_amyloid_human), '_'))[,1]
bc <- do.call(rbind, strsplit(bc, '-'))[,1]
seurat_amyloid_human$bc <- paste0(bc, '-', as.character(seurat_amyloid_human$Sample))

# match seurat metadata
ix <- match(sfe$bc, seurat_amyloid_human$bc)

# add amyloig + OC scores to the object:
sfe$Amyloglo_Score <- seurat_amyloid_human$Amyloglo_Score[ix]
sfe$OC_Score <- seurat_amyloid_human$OC_Score[ix]

# compute log scale of OC and Amyloglo scores
colData(sfe)$Amyloglo_Score_log <- log(sfe$Amyloglo_Score + 1)
colData(sfe)$OC_Score_log <- log(sfe$OC_Score + 1)

################################################################################
# geospatial stats for amyloid scores
################################################################################

feats <- c('Amyloglo_Score_log', 'OC_Score_log')

# compute Moran's I for amyloglo and OC scores
for(cur_sample in unique(seurat_amyloid_human$Sample)){
  print(cur_sample)
  sfe <- colDataUnivariate(
    sfe, type = "localmoran", features = feats,
    colGraphName = "visium", sample_id = cur_sample
  )
}

# compute Getis-Ord Gi* or amyloglo and OC scores
for(cur_sample in unique(seurat_amyloid_human$Sample)){
  print(cur_sample)
  sfe <- colDataUnivariate(
    sfe, type = "localG_perm", features = feats,
    colGraphName = "visium_B", include_self=TRUE,
    sample_id = cur_sample
  )
}

################################################################################
# geospatial stats for gene signatures
################################################################################

feats <- c(paste0(names(gene_lists), '_UCell'))

# compute Moran's I
for(cur_sample in unique(sfe$sample_id)){
  print(cur_sample)
  sfe <- colDataUnivariate(
    sfe, type = "localmoran", features = feats,
    colGraphName = "visium", sample_id = cur_sample
  )
}

# compute Getis-Ord Gi*
for(cur_sample in unique(sfe$sample_id)){
  print(cur_sample)
  sfe <- colDataUnivariate(
    sfe, type = "localG_perm", features = feats,
    colGraphName = "visium_B", include_self=TRUE,
    sample_id = cur_sample
  )
}



###############################################################
# save the SFE object
###############################################################

saveRDS(sfe, paste0(data_dir, 'ADDS_SFE.rds'))


###############################################################
# Plot the results in Seurat
###############################################################

# move the results to the seurat object:
ix <- match(as.character(sfe$bc), as.character(seurat_human$bc))
all.equal(as.character(sfe$bc[ix]), as.character(seurat_human$bc))

feats <- c(paste0(names(gene_lists), '_UCell'), 'Amyloglo_Score_log', 'OC_Score_log')

# amylo glo score Gi
for(feat in feats){

  print(feat)

  # Getis-Ord Gi*
  feat_new <- paste0(feat, '_Gi')
  test <- localResult(sfe, 'localG_perm', feat, sample_id='all')[,'localG']
  seurat_human@meta.data[[feat_new]] <- test[ix]
  seurat_human@meta.data[[feat_new]]  <- ifelse(is.na(seurat_human@meta.data[[feat_new]]), 0, seurat_human@meta.data[[feat_new]])
  seurat_human@meta.data[[feat_new]]  <- ifelse(seurat_human@meta.data[[feat_new]]  == Inf, 0, seurat_human@meta.data[[feat_new]])

  # moran's I
  feat_new <- paste0(feat, '_lmi')
  test <- localResult(sfe, 'localmoran', feat, sample_id='all')[,'Ii']
  seurat_human@meta.data[[feat_new]] <- test[ix]
  seurat_human@meta.data[[feat_new]]  <- ifelse(is.na(seurat_human@meta.data[[feat_new]]), 0, seurat_human@meta.data[[feat_new]])
  seurat_human@meta.data[[feat_new]]  <- ifelse(seurat_human@meta.data[[feat_new]]  == Inf, 0, seurat_human@meta.data[[feat_new]])
}

seurat_human$Amyloglo_Score_log <- sfe$Amyloglo_Score_log[ix]
seurat_human@meta.data[['Amyloglo_Score_log']]  <- ifelse(is.na(seurat_human@meta.data[['Amyloglo_Score_log']]), 0, seurat_human@meta.data[['Amyloglo_Score_log']])
seurat_human@meta.data[['Amyloglo_Score_log']]  <- ifelse(seurat_human@meta.data[['Amyloglo_Score_log']]  == Inf, 0, seurat_human@meta.data[['Amyloglo_Score_log']])

seurat_human$OC_Score_log <- sfe$OC_Score_log[ix]
seurat_human@meta.data[['OC_Score_log']]  <- ifelse(is.na(seurat_human@meta.data[['OC_Score_log']]), 0, seurat_human@meta.data[['OC_Score_log']])
seurat_human@meta.data[['OC_Score_log']]  <- ifelse(seurat_human@meta.data[['OC_Score_log']]  == Inf, 0, seurat_human@meta.data[['OC_Score_log']])

# save seurat object:
saveRDS(seurat_human, file=paste0(data_dir, "/ADDS_seurat_processed_geospatial.rds"))

################################################################################
# plot amyloid geospatial stats
################################################################################

# only plot samples that have amyloid scores:
amyloid_samples <- unique(seurat_amyloid_human$Sample)

for(feat in c('Amyloglo_Score_log', 'OC_Score_log')){
  feat_name <- strsplit(feat, '_')[[1]][1]
  for(modifier in c('', '_Gi', '_lmi')){

    print(feat_name)
    print(modifier)

    if(modifier == '_lmi'){
      plot_min <- 0.5
    } else{
      plot_min = 0
    }

    p <- SampleFeaturePlot(
      seurat_human,
      feature=paste0(feat, modifier),
      sample_labels = c("Diagnosis", "Sex", 'Age'),
      samples_to_plot = amyloid_samples,
      ncol = 10,
      raster=TRUE,
      plot_max = 'q99',
      plot_min = plot_min,
      colfunc = inferno,
      rev_colors=TRUE,
      dpi=800,
      combine=FALSE
    )

    p <- lapply(p, function(x){
      x + theme(
        plot.title = element_text(face='bold', size=12, vjust=-1),
        plot.margin = margin(0,0,0,0)
      ) + labs(fill = paste0(feat_name, ' ', modifier))

    })

    patch <- wrap_plots(p, ncol=8, widths=1, heights=1) + plot_layout(guides='collect')

    pdf(paste0(fig_dir, 'featureplot_', feat_name, modifier, '.pdf'), width=12, height=6)
    print(patch)
    dev.off()

  }
}


################################################################################
# plot the genescore geospatial stats
################################################################################

feats <- c(paste0(names(gene_lists), '_UCell'))

for(feat in feats){
  print(feat)
  feat_name <- strsplit(feat, '_')[[1]][1]
  for(modifier in c('', '_Gi', '_lmi')){

    print(feat_name)
    print(modifier)

    if(modifier == '_lmi'){
      plot_min <- 0.5
    } else{
      plot_min = 0
    }

    p <- SampleFeaturePlot(
      seurat_human,
      feature=paste0(feat, modifier),
      sample_labels = c("Diagnosis", "Sex", 'Age'),
      ncol = 10,
      raster=TRUE,
      plot_max = 'q99',
      plot_min = plot_min,
      colfunc = inferno,
      rev_colors=TRUE,
      dpi=800,
      combine=FALSE
    )

    p <- lapply(p, function(x){
      x + theme(
        plot.title = element_text(face='bold', size=10, vjust=-1),
        plot.margin = margin(0,0,0,0)
      ) + labs(fill = paste0(feat_name, ' ', modifier))

    })

    patch <- wrap_plots(p, ncol=10, widths=1, heights=1) + plot_layout(guides='collect')

    pdf(paste0(fig_dir, 'featureplot_', feat_name, modifier, '.pdf'), width=12, height=6)
    print(patch)
    dev.off()

  }
}

################################################################################
# test out a violin plot for one of the features:
# WOW these look bad LMAOOOOnb
################################################################################

p <- VlnPlot(
  seurat_human,
  features = 'OC_Score_log_Gi',
  group.by = 'Sample',
  #y.max = 10,
  pt.size=0,
  fill.by = 'feature'
) + NoLegend()


pdf(paste0(fig_dir, 'test_amylo_vln.pdf'), width=12, height=6)
print(p)
dev.off()


```

Amyloid Differential Gene expression analysis with Monocle3 GLM

Running this analysis with Grey + White matter together has issues, since there's
more plaques in the grey matter, it looked like the amyloid DEGs and associated
GO terms were biased towards neuronal genes up near amyloid, and glial genes
up away from the amyloid. However there were still some interesting GO results
in this more "global" analysis, with APP metabolism down and response to amyloid up.
Now going to modify to try this out in just the white matter to see what happens.


```{r eval=FALSE}

library(monocle3)
library(tictoc)

# get a list of samples that we have amyloid scores for:
amyloid_samples <- seurat_human@meta.data %>%
  group_by(Sample) %>%
  summarise(Mean = mean(Amyloglo_Score_log)) %>%
  filter(Mean != 0) %>% .$Sample %>% as.character

################################################################################
# Run DEGs with all clusters together
################################################################################

# make a new CDS with just the RNA-seq data in order to do t-DEGs:
expr_matrix  <- GetAssayData(
  subset(seurat_human, Diagnosis != 'Control' & Sample %in% amyloid_samples),
  slot='data',
)
test_genes <- rownames(expr_matrix )[rowSums(expr_matrix )!=0]

# create celldataset object
cds <- new_cell_data_set(
  expr_matrix[test_genes,],
  cell_metadata=seurat_human@meta.data[colnames(expr_matrix),]
)
print(dim(cds))

# identify OC hotspot DEGs
gene_fits <- monocle3::fit_models(
  cds,
  model_formula_str = "~OC_Score_log_Gi + Sample + nCount_Spatial",
  verbose=TRUE,
  cores=8
)

fit_coefs <- coefficient_table(gene_fits)

degs <- fit_coefs %>% filter(term == "OC_Score_log_Gi") %>%
      select(gene_id, term, p_value, q_value, test_val, std_err, estimate, normalized_effect)

# write amyloid-degs table as a csv:
write.csv(degs, file=paste0(data_dir, 'OC_Score_log_Gi_glm_DEGs.csv'), quote=FALSE, row.names=FALSE)

################################################################################
# Run DEGs with grey vs white matter separately
################################################################################

clusters <- levels(seurat_5x$annotation)
clusters <- clusters[clusters != 'unknown']

# add a new column to distinguish between white and grey matter
seurat_human$tissue_type <- ifelse(grepl('WM', as.character(seurat_human$annotation)), 'WM', 'GM')

# get the expression matrix for the selected spots
expr_matrix  <- GetAssayData(
  subset(seurat_human, Diagnosis != 'Control' & Sample %in% amyloid_samples & tissue_type == 'GM'),
  slot='data',
)
test_genes <- rownames(expr_matrix )[rowSums(expr_matrix )!=0]

# create celldataset object
cds <- new_cell_data_set(
  expr_matrix[test_genes,],
  cell_metadata=seurat_human@meta.data[colnames(expr_matrix),]
)
print(dim(cds))

# identify OC hotspot DEGs
tic()
gene_fits <- monocle3::fit_models(
  cds,
  model_formula_str = "~Amyloglo_Score_log_Gi + Sample + nCount_Spatial",
  # verbose=TRUE,
  cores=1 # got a multithreading error some times but not everytime???
)

fit_coefs <- coefficient_table(gene_fits)
elapsed_time <- toc() # ~10 minutes!

degs <- fit_coefs %>% filter(term == "Amyloglo_Score_log_Gi") %>%
      select(gene_id, term, p_value, q_value, test_val, std_err, estimate, normalized_effect)

# write amyloid-degs table as a csv:
write.csv(degs, file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'), quote=FALSE, row.names=FALSE)

degs <- read.csv(file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'))


subset(degs, q_value < 0.05) %>% dim


all.equal(as.data.frame(degs1), as.data.frame(degs2))
all.equal(as.data.frame(degs), as.data.frame(degs2))
all.equal(as.data.frame(degs), as.data.frame(degs1))


################################################################################
# Run DEGs with all clusers together
################################################################################

clusters <- levels(seurat_human$annotation)

# add a new column to distinguish between white and grey matter
seurat_human$tissue_type <- ifelse(grepl('WM', as.character(seurat_human$annotation)), 'WM', 'GM')

deg_list <- list()
for(cur_cluster in clusters){

  print(cur_cluster)

  # get the expression matrix for the selected spots
  expr_matrix  <- GetAssayData(
    subset(seurat_human, Diagnosis != 'Control' & Sample %in% amyloid_samples & annotation == cur_cluster),
    slot='data',
  )
  test_genes <- rownames(expr_matrix )[rowSums(expr_matrix )!=0]

  # create celldataset object
  cds <- new_cell_data_set(
    expr_matrix[test_genes,],
    cell_metadata=seurat_human@meta.data[colnames(expr_matrix),]
  )
  print(dim(cds))

  # identify OC hotspot DEGs
  tic()
  gene_fits <- monocle3::fit_models(
    cds,
    model_formula_str = "~Amyloglo_Score_log_Gi + Sample + nCount_Spatial",
    # verbose=TRUE,
    cores=1 # got a multithreading error some times but not everytime???
  )

  fit_coefs <- coefficient_table(gene_fits)
  elapsed_time <- toc() # ~10 minutes!

  degs <- fit_coefs %>% filter(term == "Amyloglo_Score_log_Gi") %>%
        select(gene_id, term, p_value, q_value, test_val, std_err, estimate, normalized_effect)

  degs$group <- cur_cluster
  deg_list[[cur_cluster]] <- degs

}

degs <- do.call(rbind, deg_list)

# write amyloid-degs table as a csv:
write.csv(degs, file=paste0(data_dir, 'human_Amyloglo_Score_log_Gi_DEGs.csv'), quote=FALSE, row.names=FALSE)

degs <- read.csv(file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'))

################################################################################
# Plot it as a Volcano plot?
################################################################################

library(ggrepel)

degs <- read.csv(file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'))

cp <- c("Control" = "#B8DBC5", "earlyAD" = "#E7BDE1" , "AD" = "#CF8BA3", "AD_DS" = "#9E6D7F")

# late-AD
color1 <- cp['AD']; color2 <- cp['Control']

# early-AD
color1 <- cp['earlyAD']; color2 <- cp['Control']

# AD/DS
color1 <- cp['AD_DS']; color2 <- cp['Control']

# exclude MT genes
degs <- degs[!grepl('MT-', degs$gene_id),]

# lowest non-zero value
lowest <- degs %>% subset(q_value != 0) %>% top_n(-1, wt=q_value) %>% .$q_value
degs$q_value <- ifelse(degs$q_value == 0, lowest, degs$q_value)

nlabel <- 10

cur_degs <- Reduce(rbind, lapply(unique(degs$term), function(x){
  cur <- subset(degs, term == x)

  top_thresh <- cur %>% subset(q_value <= 0.05 & normalized_effect > 0) %>% top_n(nlabel, wt=normalized_effect) %>% .$normalized_effect %>% min
  bottom_thresh <- cur %>% subset(q_value <= 0.05 & normalized_effect < 0) %>% top_n(-1*nlabel, wt=normalized_effect) %>% .$normalized_effect %>% max

  cur$anno <- ifelse(cur$q_value <= 0.05 & cur$normalized_effect >= top_thresh, cur$gene, NA)
  cur$anno <- ifelse(cur$q_value <= 0.05 & cur$normalized_effect <= bottom_thresh, cur$gene, cur$anno)
  cur$color <- ifelse(cur$q_value > 0.05, 'gray', ifelse(cur$normalized_effect > 0, color1, color2))
  cur

}))


groups <- unique(degs$term)
plot_list <- list()
for(cluster  in groups){


  print(cluster)
  plot_degs <- cur_degs %>% subset(term == cluster)

  p <- plot_degs  %>%
     ggplot(aes(x=normalized_effect, y=-log10(q_value))) +
     geom_hline(yintercept=-log10(0.05), linetype='dashed')

  # plot genes that are Nr4a2 targets
  p <- p + ggrastr::rasterise(geom_point(
    alpha=0.5,
    color=plot_degs %>% .$color
  ), dpi=500)

  p <- p +
     geom_point(
       inherit.aes=FALSE,
       data=subset(plot_degs, !is.na(anno)),
       aes(normalized_effect, -log10(q_value)),
       fill=subset(plot_degs, !is.na(anno)) %>% .$color,
       shape=21, size=3, color='black'
     ) +
     geom_text_repel(aes(label=anno), color='black', fontface='italic',  min.segment.length=0, max.overlaps=Inf) +
     xlim(-1*max(abs(plot_degs$normalized_effect))-0.1, max(abs(plot_degs$normalized_effect))+0.1) +
     ggtitle(paste0(cluster)) +
     xlab(bquote("Normalized effect size")) +
     ylab(bquote("-log"[10]~"(Adj. P-value)")) +
     theme(
       panel.border = element_rect(color='black', fill=NA, linewidth=1),
       panel.grid.major = element_blank(),
       axis.line = element_blank(),
       plot.title = element_text(hjust = 0.5, vjust=-1),
       legend.position='bottom'
     )

    plot_list[[cluster]] <- p

}

name <- 'Amyloglo_Score_log_Gi_GM'
out <- paste0(fig_dir, 'volcano_', name, '_visium.pdf')

pdf(out, width=5, height=5, useDingbats=FALSE)
plot_list[[1]]
dev.off()

# plot_list <- lapply(plot_list, function(x){
#   x + theme(
#     axis.title.x = element_blank(),
#     axis.title.y = element_blank(),
#     plot.margin = margin(0,0,0,0),
#     plot.title = element_text(vjust=-0.2)
#   )
# })
#
# pdf(out, width=12, height=12, useDingbats=FALSE)
# wrap_plots(plot_list, ncol=3)
# dev.off()


EnrichR on DEGs

```{r eval=FALSE}

################################################################################
# Run enrichr
################################################################################

degs <- read.csv(file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'))

library(enrichR)

dbs <-c('GO_Biological_Process_2021','GO_Cellular_Component_2021','GO_Molecular_Function_2021', 'WikiPathway_2021_Mouse', 'KEGG_2021_Mouse')

# get genes that are up / down
cur_up <- subset(degs, q_value < 0.05 & normalized_effect >= logfc_thresh) %>% .$gene_id
cur_down <- subset(degs, q_value < 0.05 & normalized_effect <= -logfc_thresh) %>% .$gene_id

# setup input list for enrichr loop
input_list <- list(
  up_amyloid = cur_up,
  down_amyloid = cur_down
)

# run enrichr and combine outputs
enriched_df <- do.call(rbind, lapply(names(input_list), function(x){
  print(x)
  print(length(input_list[[x]]))
  if(length(input_list[[x]]) > 0){
    cur_enrich <- enrichr(input_list[[x]], dbs)
  } else{return(data.frame())}
  cur_df <- do.call(rbind, lapply(dbs, function(cur_db){
    df <- cur_enrich[[cur_db]]
    if(nrow(df) > 1){
      df$degs <- x;
      # df$group <- cur_group;
      df$db <- cur_db
    }
    else{df <- data.frame()}
    print(dim(df))
    df
  }))
}))

enriched_df %>% subset(P.value < 0.05) %>%
write.table(file=paste0(data_dir, 'test_amyloglo_GO_terms.tsv'), quote=FALSE, row.names=FALSE, sep='\t')



################################################################################
# plot selected go terms as a bar plot
################################################################################

# helper function to wrap text
wrapText <- function(x, len) {
    sapply(x, function(y) paste(strwrap(y, len), collapse = "\n"), USE.NAMES = FALSE)
}

selected_terms <- read.delim('data/human_Amyloglo_GO_terms_selected.txt', sep='\t', header=1)

# remove GO Term ID
selected_terms$Term <- str_replace(selected_terms$Term, " \\s*\\([^\\)]+\\)", "")

selected_terms <- selected_terms %>%
  group_by(module) %>%
  arrange(Combined.Score)

selected_terms$wrap <- wrapText(selected_terms$Term, 35)

p <- selected_terms  %>%
  ggplot(aes(x=log(Combined.Score), y=reorder(Term, Combined.Score)))+
  geom_bar(stat='identity', position='identity', color='white', fill='grey65') +
  geom_text(aes(label=Term), x=.1, color='black', size=3.5, hjust='left') +
  ylab('') + xlab(bquote("log"[10]~"(Enrich)")) +
  scale_x_continuous(expand = c(0, 0), limits = c(0, NA)) +
  theme(
    panel.grid.major=element_blank(),
    panel.grid.minor=element_blank(),
    legend.title = element_blank(),
    axis.ticks.y=element_blank(),
    axis.text.y=element_blank(),
    axis.line.y=element_blank(),
    plot.title = element_text(hjust = 0.5)
  )

pdf(paste0(fig_dir, 'human_amyloglo_selected_GO_terms.pdf'), width= 3, height=4 , useDingbats=FALSE)
p
dev.off()

```


Overlap with known gene lists

```{r eval=FALSE}

library(GeneOverlap)


degs <- read.csv(file=paste0(data_dir, 'Amyloglo_Score_log_Gi_glm_GM_DEGs.csv'))

# upregulated genes :
amylo_genes <- degs %>% subset(q_value < 0.05 & normalized_effect > 0) %>% .$gene
amylo_genes <- amylo_genes[!grepl('MT-', amylo_genes)]


overlap_df <- do.call(rbind, lapply(names(gene_lists), function(cur_list){

  cur_genes <- gene_lists[[cur_list]]

  cur_overlap <- testGeneOverlap(newGeneOverlap(
      cur_genes,
      amylo_genes,
      genome.size=nrow(seurat_human)
  ))

  cur_overlap <- data.frame(
    'odds.ratio' = cur_overlap@odds.ratio,
    'pval' = cur_overlap@pval,
    'Jaccard' = cur_overlap@Jaccard,
    'size_intersection' = length(cur_overlap@intersection),
    'list' = cur_list
  )

  cur_overlap

})) %>% as.data.frame()

overlap_df <- overlap_df %>% mutate(fdr=p.adjust(pval, method='fdr'))


################################################################################
# Plot as a lollipop
################################################################################

overlap_df$shape <- ifelse(overlap_df$fdr < 0.05, 21, 4)
overlap_df <- overlap_df %>% arrange(odds.ratio, descending=TRUE)
overlap_df$module <- factor(as.character(overlap_df$module), levels=as.character(overlap_df$module))

mod_colors <- dplyr::select(modules, c(module, color)) %>%
  distinct
cp <- mod_colors$color; names(cp) <- mod_colors$module

p <- overlap_df %>%
  ggplot(aes(y=module, x=odds.ratio, size= size_intersection, color=module)) +
  geom_segment(aes(y=module, yend=module, x=0, xend=odds.ratio), size=0.5, color='grey') +
  geom_point() +
  geom_point(shape=overlap_df$shape, color='black', fill=NA) +
  scale_color_manual(values=cp, guide='none') +
  ylab('') + xlab("Odds ratio") +
  scale_x_continuous(breaks = c(0, 1, 2,3)) +
  labs(size='Size\nintersection') +
  ggtitle('Overlap with SFARI genes') +

  theme(
    panel.border = element_rect(size=1, color='black', fill=NA),
    axis.line.y = element_blank(),
    axis.line.x = element_blank(),
    plot.title = element_text(hjust=0.5, face='plain')
  )

pdf(paste0(fig_dir, 'sfari_overlap2.pdf'), width=4, height=3.5)
p
dev.off()


```

```{r eval=FALSE}

################################################################################
# Overlap with other gene lists:
#
# * ST marker genes
# * SC marker genes
# * AD vs Control DEGs ST & SC
# * ADDS vs Control DEGs ST & SC
# * Known gene signatures (DAM, DAA, DOL, PIGs, GWAS)
################################################################################


intersect(pigs, cur_up)
intersect(pigs, cur_down)

intersect(oligs, cur_up)
intersect(oligs, cur_down)

```
