
```{r eval=FALSE}

library(Seurat)
library(tidyverse)
library(cowplot)
library(patchwork)
library(ggrepel)
library(viridis)
library(magrittr)
library(hdWGCNA)
theme_set(theme_cowplot())

source('/dfs7/swaruplab/smorabit/analysis/scWGCNA/bin/spatial_functions.R')


setwd("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/scDRS/")

fig_dir <- 'figures/'
data_dir <- 'data/'

# load ADDS + AD integrated single-cell dataset:
seurat_human <- readRDS("/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/voyager/data/ADDS_seurat_processed_geospatial.rds")


human_cp <- c(
      "L1" = "#8B3D5A", "L2-3" = "#E7BDE1", "L3-4" = "#E6A4CD",
      "L3-4-5" = "#CF8BA3", "L5-6" = "#9E6D7F", "L6b" = "#CDAEB9", "WM" = "#62A7D7" )

color_df <- data.frame(
  colour = as.character(human_cp),
  group = names(human_cp)
)
color_df$group <- factor(
  as.character(color_df$group),
    levels=names(human_cp)
)
color_df <- arrange(color_df, group)


```

Combine the scDRS output tables for each dataset 

```{r eval=FALSE}


# get the file paths for the output files
scdrs_output_dir <- paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/scDRS/data/outputs/')
output_files <- dir(scdrs_output_dir)
output_files <- output_files[!grepl('full', output_files)]
traits <- gsub(".score.gz", '', output_files)
traits <- gsub("PASS_", '', traits)
output_files <- paste0(scdrs_output_dir, output_files)

# combine the individual tables
scdrs_df <- do.call(rbind, lapply(1:length(traits), function(i){
    cur_df <- read_delim(output_files[i],  delim='\t')
    names(cur_df)[1] <- 'barcode'
    cur_df$trait <- traits[i]
    cur_df
}))

# write the output file:
write.csv(scdrs_df, quote=FALSE, row.names=FALSE, file=paste0(data_dir,'ADDS_visium_scDRS_results.csv'))


```

Plot the AD scDRS score as a spatial featureplot 

```{r eval=FALSE}

# load the AD/DS results
scdrs_df <- read_csv(file=paste0(data_dir, 'ADDS_visium_scDRS_results.csv'))

# subset for one trait (AD)
cur_df <- subset(scdrs_df, trait == 'Alzheimers_Jansen2019')


cur_seurat <- subset(seurat_human, bc %in% cur_df$barcode)

ix <- match(cur_df$barcode, cur_seurat$bc)
all.equal(as.character(cur_df$barcode[ix]), as.character(cur_seurat$bc))

cur_seurat$scdrs_score <- cur_df[ix,]$norm_score
cur_seurat$scdrs_zscore <- cur_df[ix,]$zscore
cur_seurat$scdrs_pval <- cur_df[ix,]$pval
cur_seurat$scdrs_score_signif <- ifelse(cur_seurat$scdrs_pval < 0.05, cur_seurat$scdrs_score, 0)

# representative samples
human_rep_samples <- c(
  'Dec_13_2021_Human5', 'Dec_20_2021_Human1',
  'Dec_13_2021_Human6', 'Oct_2021_6',
  'Dec_13_2021_Human3', 'Dec_13_2021_Human7',
  'Nov_24_2021_VisiumHuman_12', 'Dec_13_2021_Human8'
)


library(RColorBrewer)

p <- SampleFeaturePlot(
    cur_seurat,
    feature='scdrs_score',
    sample_labels = c("Sex", 'Diagnosis'),
    sample_col = "Sample",
    samples_to_plot = human_rep_samples,
    ncol = 8,
    raster=TRUE,
    plot_max = 'q99',
    plot_min = 0,
    colfunc = inferno,
    rev_colors=TRUE,
   # colfunc=colorRampPalette(c('grey90', cur_color), bias=2),
   #rev_colors=FALSE,
    dpi=300,
    combine=FALSE
)

# p <- lapply(p, function(x){
#     x + scale_color_gradient2()
# })

patch <- wrap_plots(p, ncol=8) + plot_layout(guides='collect')

pdf(paste0(fig_dir, 'ADDS_visium_scDRS_AD_featureplot.pdf'), width=16, height=4)
patch
dev.off()


```


Compute the celltype / disease associations for the whole dataset

```{r eval=FALSE}

all(seurat_human$bc %in% scdrs_df$barcode)
all(colnames(seurat_human) %in% scdrs_df$barcode)

seurat_human$bc <- colnames(seurat_human)
seurat_human$region <- ifelse(grepl('WM', as.character(seurat_human$annotation)), 'WM', as.character(seurat_human$annotation))

scdrs_output_dir <- paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/scDRS/data/outputs/')
output_files <- dir(scdrs_output_dir)
output_files <- output_files[grepl('full', output_files)]
traits <- gsub(".score.gz", '', output_files)
traits <- gsub("PASS_", '', traits)
output_files <- paste0(scdrs_output_dir, output_files)


assoc_df <- data.frame()
for(i in 1:length(traits)){
    tic()
    out <- TraitGroupAssociations(
        seurat_obj = seurat_human,
        scdrs_file = output_files[i],
        group.by = 'region',
        percentile = 0.95
    )
    out$trait <- traits[i]
    time_elapsed <- toc()
    print(time_elapsed)
    assoc_df <- rbind(assoc_df, out)
}

# compute the FDR
assoc_df$mc_fdr <- p.adjust(assoc_df$mc_p, method='fdr')

write.csv(assoc_df, file=paste0(data_dir, 'ADDS_visium_scDRS_celltype_associations.csv'))

#-------------------------------------------------------------------------------------#
# quick and dirty of plot the results as a heatmap 
#-------------------------------------------------------------------------------------#

p <- ggplot(assoc_df, aes(x=group, y=trait, fill=prop_p_0.05)) + 
    geom_tile() + 
    geom_point(
        inherit.aes=FALSE,
        data = subset(assoc_df, mc_fdr < 0.05),
        aes(x=group, y=trait),
        color='black', 
        shape=0,
        size=4,
        fill=NA
    ) + 
    coord_equal() + 
    scale_fill_steps(low='white', high='red') + 
    RotatedAxis()

pdf(paste0(fig_dir, 'ADDS_visium_scdrs_heatmap.pdf'), width=12, height=15)
p
dev.off()



```





Compute the celltype / disease associations separately for each condition 

```{r eval=FALSE}

all(seurat_human$bc %in% scdrs_df$barcode)
all(colnames(seurat_human) %in% scdrs_df$barcode)

seurat_human$bc <- colnames(seurat_human)
seurat_human$region <- ifelse(grepl('WM', as.character(seurat_human$annotation)), 'WM', as.character(seurat_human$annotation))

# get a list of scDRS output files:
scdrs_output_dir <- paste0('/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/scDRS/data/outputs/')
output_files <- dir(scdrs_output_dir)
output_files <- output_files[grepl('full', output_files)]
traits <- gsub(".score.gz", '', output_files)
traits <- gsub("PASS_", '', traits)
output_files <- paste0(scdrs_output_dir, output_files)

# list of different conditions:
conditions <- levels(seurat_human$Diagnosis)


assoc_df <- data.frame()
for(j in 1:length(conditions)){
    cur_condition <- conditions[j]
    print(cur_condition)
    cur_seurat <- subset(seurat_human, Diagnosis == cur_condition)
    for(i in 1:length(traits)){
        out <- TraitGroupAssociations(
            seurat_obj = cur_seurat,
            scdrs_file = output_files[i],
            group.by = 'region',
            percentile = 0.95
        )
        out$trait <- traits[i]
        out$condition <- cur_condition
        assoc_df <- rbind(assoc_df, out) 
    }
}

# compute the FDR
assoc_df$mc_fdr <- p.adjust(assoc_df$mc_p, method='fdr')
write.csv(assoc_df, file=paste0(data_dir, 'ADDS_visium_diagnosis_scDRS_celltype_associations.csv'))

assoc_df %<>% 
    group_by(condition) %>% 
    mutate(fdr = p.adjust(mc_p, method='fdr')) %>% 
    ungroup

#-------------------------------------------------------------------------------------#
# quick and dirty of plot the results as a heatmap 
#-------------------------------------------------------------------------------------#

p <- ggplot(assoc_df, aes(x=group, y=trait, color=prop_p_0.05, size=-log10(fdr))) + 
    geom_point() + 
    geom_point(
        inherit.aes=FALSE,
        data = subset(assoc_df, fdr <= 0.05),
        aes(x=group, y=trait, size=-log10(fdr)),
        color='black', 
        shape=1,
        #size=-log10(fdr),
        fill=NA
    ) + 
    coord_equal() + 
    scale_color_steps(low='white', high='red') + 
    RotatedAxis()

pdf(paste0(fig_dir, 'ADDS_visium_scdrs_heatmap.pdf'), width=20, height=15)
p + facet_wrap(~condition, ncol=4)
dev.off()

assoc_df %>% subset(trait == 'UKB_460K.body_HEIGHTz.full')


p <- ggplot(assoc_df, aes(x=group, y=trait, color=prop_p_0.05, size=-log10(fdr))) + 
    geom_point() + 
    geom_point(
        inherit.aes=FALSE,
        data = subset(assoc_df, fdr <= 0.05),
        aes(x=group, y=trait, size=-log10(fdr)),
        color='black', 
        shape=1,
        #size=-log10(fdr),
        fill=NA
    ) + 
    coord_equal() + 
    scale_color_steps(low='white', high='red') + 
    RotatedAxis()

pdf(paste0(fig_dir, 'ADDS_visium_scdrs_heatmap_ratio.pdf'), width=20, height=15)
p + facet_wrap(~condition, ncol=4)
dev.off()


```


Ordering of the traits in the heatmap 

```{r eval=FALSE}


blood_immune_traits <- c(
    'CD_deLange2017', # chron's disease
    'Celiac',
    'IBD_deLange2017',
    'Lupus',
    'Multiple_sclerosis',
    'Rheumatoid_Arthritis',
    'Primary_biliary_cirrhosis',
    'UC_deLange2017',
    'Type_1_Diabetes',
    'EOSINOPHIL_COUNT',
    'LYMPHOCYTE_COUNT',
    'MEAN_CORPUSCULAR_HEMOGLOBIN',
    'MONOCYTE_COUNT',
    'PLATELET_COUNT',
    'RBC_DISTRIB_WIDTH',
    'RED_COUNT',
    'WHITE_COUNT',
    'AID_ALL', #AID = autoimmune disease
    'ALLERGY_ECZEMA_DIAGNOSED',
    'ASTHMA_DIAGNOSED',
    'HYPOTHYROIDISM_SELF_REP',
    'RESPIRATORY_ENT'
)

brain_traits <- c(
    'ADHD_Demontis2018',
    'Alzheimers_Jansen2019',
    'BIP_Mullins2021',
    'DrinksPerWeek_Liu2019',
    'GeneralRiskTolerance_KarlssonLinner2019',
    'Insomnia_Jansen2019',
    'Intelligence_SavageJansen2018',
    'MDD_Howard2019',
    'ReactionTime_Davies2018',
    'SWB',
    'Schizophrenia_Pardinas2018',
    'SleepDuration_Dashti2019',
    'VerbalNumericReasoning_Davies2018',
    'EDU_COLLEGE',
    'EDU_YEARS',
    'NEUROTICISM',
    'MORNINGPERSON',
    'BMIz',
    'SMOKING_STATUS',
    'NumberChildrenEverBorn_Pooled',
    'Worry_Nagel2018'
)

other_diseases <- c(
    'Type_2_Diabetes',
    'AtrialFibrillation_Nielsen2018',
    'Coronary_Artery_Disease',
    'SYSTOLICadjMEDz',
    'DIASTOLICadjMEDz',
    'CARDIOVASCULAR',
    'HYPERTENSION_DIAGNOSED',
    'FastingGlucose_Manning',
    'AlanineAminotransferase',
    'AlkalinePhosphatase',
    'Cholesterol',
    'Glucose',
    'HDLcholesterol',
    'HbA1c',
    'LDLdirect',
    'SHBG',
    'Testosterone_Male',
    'TotalBilirubin',
    'TotalProtein',
    'Triglycerides',
    'HEEL_TSCOREz',
    'BALDING1',
    'HEIGHTz',
    'WHRadjBMIz',
    'BREAST',
    'BASAL_METABOLIC_RATEz',
    'FEV1FVCzSMOKE',
    'FVCzSMOKE',
    'HAIR',
    'MENARCHE_AGE',
    'MENOPAUSE_AGE'
)



traits_order <- c(brain_traits, blood_immune_traits, other_diseases)

```

```{r eval=FALSE}

# re-load  file 
combined_assoc_df <- read.csv(file=paste0(data_dir, 'ADDS_visium_scDRS_celltype_associations.csv'))
combined_assoc_df$condition <- 'All'
combined_assoc_df$mc_fdr <- p.adjust(combined_assoc_df$mc_p, method='fdr')

# re-load  file 
assoc_df <- read.csv(file=paste0(data_dir, 'ADDS_visium_diagnosis_scDRS_celltype_associations.csv'))
assoc_df$mc_fdr <- p.adjust(assoc_df$mc_p, method='fdr')

# combine
assoc_df <- rbind(assoc_df, combined_assoc_df)
#assoc_df$group <- factor(as.character(assoc_df$group), levels=levels(seurat_human$region))
#assoc_df$group <- droplevels(assoc_df$group)

# rename for convenience
assoc_df$trait <- gsub('UKB_460K.biochemistry_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.blood_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.bmd_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.body_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.bp_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.cancer_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.cov_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.disease_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.impedance_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.lung_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.mental_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.other_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.pigment_', '', assoc_df$trait)
assoc_df$trait <- gsub('UKB_460K.repro_', '', assoc_df$trait)
assoc_df$trait <- gsub('.full', '', assoc_df$trait)

traits <- unique(assoc_df$trait)

# add trait type:
assoc_df$trait_type <- ifelse(assoc_df$trait %in% brain_traits, 'brain', 'other')
assoc_df$trait_type <- ifelse(assoc_df$trait %in% blood_immune_traits, 'blood/immune', assoc_df$trait)

#table(assoc_df$condition, assoc_df$trait)

#------------------------------------------------------------------------#
# different plots for each condition 
#------------------------------------------------------------------------#

plot_groups <- unique(assoc_df$condition)
plot_groups <- c('Control', 'earlyAD', 'AD', 'AD_DS', 'All')

plot_list <- list()
i <- 1
for(cur_group in plot_groups){

    plot_df <- subset(assoc_df, condition == cur_group)
    plot_df$trait <- factor(as.character(plot_df$trait), levels=rev(traits_order))

    p <- ggplot(plot_df, aes(x=group, y=trait, color=prop_p_0.05, size=-log10(mc_fdr))) + 
        geom_point() + 
        geom_point(
            inherit.aes=FALSE,
            data = subset(plot_df, mc_fdr <= 0.05),
            aes(x=group, y=trait, size=-log10(mc_fdr)),
            color='black', 
            shape=1,
            fill=NA
        ) + 
        coord_equal() + 
        scale_color_steps(
            low='white', high=as.character(dx_cp[cur_group]),
           # breaks=seq(0:10)/10
        ) + 
       # scale_size_continuous(range = c(1, 3 * max(-log10(assoc_df$mc_fdr)))) +
        xlab('') + ylab('') +
        ggtitle(cur_group) +
        theme(
            axis.line.x = element_blank(),
            axis.line.y = element_blank(),
            panel.border = element_rect(linewidth=1, color='black', fill=NA),
            axis.text.x = element_blank(),
            axis.title.x = element_blank(),
            axis.ticks.x = element_blank(),
            plot.title = element_text(hjust=0.5),
            panel.grid = element_line(size=0.25, color='lightgrey'),
            plot.margin = margin(c(0,0,0,0))
        ) 

    if(cur_group != plot_groups[1]){
        p <- p + theme( axis.text.y = element_blank())
    }

    # # make the colorbar as its own heatmap
    color_df$var <- 1
    cp <- color_df$colour; names(cp) <- color_df$group
    colorbar <- color_df %>% 
        subset(group %in% assoc_df$group) %>% 
        mutate(group = droplevels(group)) %>%
        ggplot(aes(x=group, y=var, fill=group)) +
        geom_tile() +
        scale_fill_manual(values=cp) +
        coord_equal() +
        NoLegend() + RotatedAxis() +
        theme(
            plot.title=element_blank(),
            axis.line=element_blank(),
            axis.ticks.y =element_blank(),
            axis.text.y = element_blank(),
            axis.title = element_blank(),
            plot.margin=margin(0,0,0,0),
        )


    patch <- p  / colorbar
    plot_list[[cur_group]] <- patch

}

pdf(paste0(fig_dir, 'ADDS_visium_scdrs_heatmap_ordered.pdf'), width=14, height=18)
wrap_plots(plot_list, ncol=5) + plot_layout(guides='collect')
dev.off()


#------------------------------------------------------------------------#
# Make a dot plot just for the AD trait
#------------------------------------------------------------------------#



cur_trait <- 'Alzheimers_Jansen2019'
plot_groups <- c('Control', 'earlyAD', 'AD', 'AD_DS')

# get the info for this trait 
plot_df <- subset(assoc_df, trait == cur_trait & condition != 'All')
plot_df$condition <- factor(
    as.character(plot_df$condition), 
    levels=plot_groups
)
plot_df$group <- fct_rev(as.factor(as.character(plot_df$group)))

p <- ggplot(plot_df, aes(x=condition, y=group, color=prop_p_0.05, size=-log10(mc_fdr))) + 
    geom_point() + 
    geom_point(
        inherit.aes=FALSE,
        data = subset(plot_df, mc_fdr <= 0.05),
        aes(x=condition, y=group, size=-log10(mc_fdr)),
        color='black', 
        shape=1,
        fill=NA
    ) + 
    coord_equal() + 
    scale_color_steps(
        low='white', high='red',
        # breaks=seq(0:10)/10
    ) + 
    scale_size(limits = c(0, 2.5)) +
    xlab('') + ylab('') +
   # ggtitle(cur_group) +
    theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(linewidth=1, color='black', fill=NA),
       # axis.text.x = element_blank(),
        axis.title.x = element_blank(),
       # axis.ticks.x = element_blank(),
        plot.title = element_text(hjust=0.5),
        panel.grid = element_line(size=0.25, color='lightgrey'),
        plot.margin = margin(c(0,0,0,0))
    ) + RotatedAxis()



pdf(paste0(fig_dir, 'human_visium_AD_scDRS_dotplot.pdf'), width=4, height=3)
p 
dev.off()


#------------------------------------------------------------------------#
# Make a line plot just for the AD trait
#------------------------------------------------------------------------#

plot_groups <- c('Control', 'earlyAD', 'AD', 'AD_DS')
cur_trait <- 'Alzheimers_Jansen2019'

# get the info for this trait 
plot_df <- subset(assoc_df, trait == cur_trait & condition != 'All')
plot_df$condition <- factor(as.character(plot_df$condition), levels=plot_groups)

# 
p <- ggplot(plot_df, aes(x = condition, y=-log10(mc_fdr), color=group, size=prop_p_0.05/2, group=group)) + 
    geom_rect(
          data = plot_df[1,],
          inherit.aes=FALSE,
    aes(xmin=-Inf, xmax=Inf, ymin=-Inf, ymax=-log10(0.05)), fill='grey87', alpha=0.8, color=NA) +
    geom_hline(yintercept=-log10(0.05), linetype='dashed', color='black') +
    geom_line(alpha=0.75, linewidth=1) + 
    geom_point() +
    geom_point(
        inherit.aes=FALSE,
        data = subset(plot_df, mc_fdr <= 0.05),
        aes(x = condition, y=-log10(mc_fdr), fill=group, size=prop_p_0.05/2, group=group),
        color='black',
        shape=21
    ) + 
    geom_text_repel(
        data = subset(plot_df, mc_fdr <= 0.05),
        aes(x=plot_groups[length(plot_groups)], y = -log10(mc_fdr) , label=group),
        color='black',
        nudge_x = 0.5,
        direction = "y",
        hjust = "left",
        min.segment.length=0,
        max.overlaps=Inf
      ) +
    scale_color_manual(values=human_cp) + 
    scale_fill_manual(values=human_cp) + 
    theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(linewidth=1, color='black', fill=NA),
        plot.title = element_text(hjust=0.5),
        panel.grid = element_line(size=0.25, color='white'),
    ) + RotatedAxis()


pdf(paste0(fig_dir, 'scDRS_human_Visium_AD.pdf'), width=4, height=4)
p
dev.off()


```


Correlation of scDRS and amyloid hotspot 

Results are very low correlations!

```{r eval=FALSE}


# load the AD/DS results
scdrs_df <- read_csv(file=paste0(data_dir, 'ADDS_visium_scDRS_results.csv'))

# subset for one trait (AD)
cur_df <- subset(scdrs_df, trait == 'Alzheimers_Jansen2019')

ix <- match(cur_df$barcode, names(seurat_human$bc))
all.equal(as.character(cur_df$barcode[ix]), names(seurat_human$bc))

seurat_human$scdrs_score <- cur_df[ix,]$norm_score
seurat_human$scdrs_zscore <- cur_df[ix,]$zscore
seurat_human$scdrs_pval <- cur_df[ix,]$pval
seurat_human$scdrs_score_signif <- ifelse(seurat_human$scdrs_pval < 0.05, seurat_human$scdrs_score, 0)



samples <- unique(seurat_human$Sample)

amyloid_col <- 'DAM_UCell'
amyloid_col <- 'Amyloglo_Score_log'
amyloid_col <- 'OC_Score_log'
scdrs_col <- 'scdrs_zscore'
cor_df <- data.frame()

for(cur_sample in samples){

    print(cur_sample)
    cur_meta <- subset(seurat_human@meta.data, Sample == cur_sample)
    cur_cor <- cor.test(
      cur_meta[[amyloid_col]],
      cur_meta[[scdrs_col]]
    )

    cur_df <- data.frame(
        sample = cur_sample,
        cor = as.numeric(cur_cor$estimate),
        pval = as.numeric(cur_cor$p.value)
    )
    cor_df <- rbind(cor_df, cur_df)

}

cor_df <- subset(cor_df, ! is.na(cor))
quantile(cor_df$cor)




subset(seurat_human@meta.data, scdrs_pval <= 0.05) %>% .$scdrs_score %>% min


# plot the scDRS score
p <- SampleFeaturePlot(
    seurat_human,
    feature='scdrs_zscore',
    sample_labels = c("Diagnosis", "Sex", 'Age'),
    ncol = 10,
    raster=TRUE,
    plot_max = 'q99',
    plot_min = 1.5,
    colfunc = inferno,
    rev_colors=TRUE,
    dpi=800,
    combine=FALSE
)

p <- lapply(p, function(x){
    x + theme(
        plot.title = element_text(face='bold', size=10, vjust=-1),
        plot.margin = margin(0,0,0,0)
    ) 
   # scale_fill_gradient2()
})


patch <- wrap_plots(p, ncol=10, widths=1, heights=1) + plot_layout(guides='collect')

pdf(paste0(fig_dir, 'featureplot_scDRS_AD.pdf'), width=12, height=6)
print(patch)
dev.off()



```

randomly check for GWAS genes 

```{r eval=FALSE}

modules <- GetModules(seurat_human)


subset(modules, gene_name == 'BIN1') # M1


subset(modules, gene_name == 'SORL1') # M6
subset(modules, gene_name == 'ADAM17') # M6
subset(modules, gene_name == 'APOE') # M6
subset(modules, gene_name == 'APP') # M6

subset(modules, gene_name == 'ADAMTS1') # M11
subset(modules, gene_name == 'CLU') # M11


subset(modules, gene_name == 'TREM2') 
subset(modules, gene_name == 'CD2AP') # M6
subset(modules, gene_name == 'EPDR1') # M4 
subset(modules, gene_name == 'PTK2B') # M3
subset(modules, gene_name == 'FERMT2') # M11
subset(modules, gene_name == 'SLC24A4') # M2
subset(modules, gene_name == 'SPPL2A') # M2
subset(modules, gene_name == 'SPPL2A') # M2
subset(modules, gene_name == 'MINDY2') # M12
subset(modules, gene_name == 'APH1B') # M2
subset(modules, gene_name == 'BCKDK') # M7
subset(modules, gene_name == 'IL34') # M9
subset(modules, gene_name == 'PLCG2') # M3
subset(modules, gene_name == 'TSPOAP1') # M13
subset(modules, gene_name == 'ABCA7') # M6
subset(modules, gene_name == 'SORT1') # M1
subset(modules, gene_name == 'EIF2AK2') # M6
subset(modules, gene_name == 'CEBPZOS') # M2
subset(modules, gene_name == 'NCK2') # M6
subset(modules, gene_name == 'ICA1L') # M2
subset(modules, gene_name == 'DGKQ') # M6
subset(modules, gene_name == 'COX7C') # M14
subset(modules, gene_name == 'TNIP1') # M6
subset(modules, gene_name == 'HS3ST5') # M3
subset(modules, gene_name == 'UMAD1') # M2
subset(modules, gene_name == 'ICA1') # M2
subset(modules, gene_name == 'TMEM106B') # M2
subset(modules, gene_name == 'JAZF1') # M3
subset(modules, gene_name == 'EGFR') # M6
subset(modules, gene_name == 'CTSB') # M7
subset(modules, gene_name == 'SHARPIN') # M6
subset(modules, gene_name == 'CCDC6') # M4
subset(modules, gene_name == 'MAF') # M2
subset(modules, gene_name == 'IQCD') # M




cur_mod_color <- subset(modules, module == 'M11') %>% .$color %>% unique

seurat_human <- RunModuleUMAP(
  seurat_human,
  n_hubs = 10,
  genes_use = modules %>% subset(module == 'M11') %>% .$gene_name,
  n_neighbors=25,
  min_dist=0.1,
  spread=1,
  target_weight=0.5
)

# get the hub gene UMAP table from the seurat object
umap_df <- GetModuleUMAP(
  seurat_human
)

# plot with ggplot
p <- ggplot(umap_df, aes(x=UMAP1, y=UMAP2)) +
  geom_point(
   color=umap_df$color,
   size=umap_df$kME*2
  ) + 
  geom_text_repel(
    aes(label=gene), size=2, max.overlaps=Inf
  )

pdf(paste0(fig_dir, 'hubgene_umap_M11.pdf'), width=8, height=8)
p
dev.off()

  
TOM <- GetTOM(seurat_human)

selected_genes <- modules %>% subset(module == 'M11') %>% .$gene_name
cur_TOM <- TOM[selected_genes,selected_genes]

cur_network <- reshape2::melt(cur_TOM)# %>% 
   # dplyr::rename()

# only keep the 5% strongest edges:
edge_cutoff <- as.numeric(quantile(cur_network$value, 0.90))
cur_network <- subset(cur_network, value >= edge_cutoff)

library(ggraph)
library(tidygraph)
library(igraph)

graph <- as_tbl_graph(cur_network) %>% 
  activate(nodes) %>% 
  mutate(degree  = centrality_degree())  


# make the layout table using the umap coords:
umap_layout <- umap_df[names(V(graph)),] %>% dplyr::rename(c(x=UMAP1, y = UMAP2, name=gene))
rownames(umap_layout) <- 1:nrow(umap_layout)
lay <- create_layout(graph, umap_layout)


p <- ggraph(lay) + 
  ggrastr::rasterise(
    geom_point(inherit.aes=FALSE, data=umap_df, aes(x=UMAP1, y=UMAP2), color=umap_df$color, alpha=0.1, size=3),
    dpi=500
   ) +
  geom_edge_fan(aes(color=value, alpha=abs(value))) + 
  geom_node_point(aes(fill=module, size=degree), color=cur_mod_color) +
  geom_node_text(aes(label=name), color='black', repel=TRUE, max.overlaps=Inf, fontface='italic', size=2) +
  scale_edge_colour_gradient(high=cur_mod_color, low='white')

pdf(paste0(fig_dir,'test_network4.pdf'), width=10, height=9)
print(p)
dev.off()




```


Correlation with hdWGCNA MEs 

```{r eval=FALSE}

# load the seurat object with hdWGCNA info (takes a long time to load)
seurat_human <- readRDS(file="/dfs7/swaruplab/smorabit/analysis/ADDS_2021/visium/human/hdWGCNA/data/ADDS_seurat_processed_annotated_hdWGCNA_all.rds")

seurat_human$anno_DMEs <- ifelse(
    grepl('WM', as.character(seurat_human$annotation)), 
    'WM', as.character(seurat_human$annotation)
)
seurat_human$anno_DMEs <- gsub('-', '', seurat_human$anno_DMEs)

# load the AD/DS results
scdrs_df <- read_csv(file=paste0(data_dir, 'ADDS_visium_scDRS_results.csv'))

# subset for one trait (AD)
cur_df <- subset(scdrs_df, trait == 'Alzheimers_Jansen2019')

ix <- match(cur_df$barcode, colnames(seurat_human))
all.equal(as.character(cur_df$barcode[ix]), colnames(seurat_human))

seurat_human$scdrs_score <- cur_df[ix,]$norm_score
seurat_human$scdrs_zscore <- cur_df[ix,]$zscore
seurat_human$scdrs_pval <- cur_df[ix,]$pval
seurat_human$scdrs_score_signif <- ifelse(seurat_human$scdrs_pval < 0.05, seurat_human$scdrs_score, 0)


traits <- c('scdrs_score', 'scdrs_zscore')


conditions <- unique(seurat_human$Diagnosis)
wgcna_groups <- c('L1', 'L2-3', 'L3-4', 'L3-4-5', 'L5-6', 'L6b', 'WM', 'consensus')

seurat_human$Condition_Group <- paste0(
    as.character(seurat_human$Diagnosis),
    '-',
    as.character(seurat_human$anno_DMEs)
)

plot_df <- data.frame()
for(cur_condition in conditions){
    for(cur_wgcna in wgcna_groups){
        
        print(cur_wgcna)
        print(cur_condition)        

        seurat_human <- SetActiveWGCNA(seurat_human, cur_wgcna)
        seurat_human <- ModuleTraitCorrelation(
            seurat_human,
            traits = traits,
            group.by = 'anno_DMEs',
            subset_by = 'Diagnosis',
            subset_groups = cur_condition
        )


        mt_cor <- GetModuleTraitCorrelation(seurat_human)
        names(mt_cor)

        cor_df <- do.call(rbind, lapply(names(mt_cor$cor), function(x){
            cur_cor <- mt_cor$cor[[x]]
            cur_fdr <- mt_cor$fdr[[x]]
            cur_cor_df <- reshape2::melt(cur_cor)
            cur_fdr_df <- reshape2::melt(cur_fdr)
            cur_cor_df$fdr <- cur_fdr_df$value
            cur_cor_df$group <- x
            cur_cor_df
        }))
        cor_df %<>% dplyr::rename(c(trait = Var1, module = Var2, cor=value))

        cor_df$Diagnosis <- cur_condition 
        cor_df$wgcna <- cur_wgcna

        plot_df <- rbind(plot_df, cor_df)

    }
}

write.csv(plot_df, file=paste0(data_dir, 'scDRS_module_correlation.csv'), quote=FALSE)

plot_df <- read.csv(file=paste0(data_dir, 'scDRS_module_correlation.csv'))


tmp <- subset(plot_df, wgcna == 'consensus')
tmp2 <- subset(plot_df, wgcna != 'consensus')
ix <- do.call(rbind, strsplit(tmp2$module, '-'))[,1]
tmp2 <- tmp2[ix == gsub('-', '', tmp2$group),]

plot_df <- rbind(tmp, tmp2)

plot_df$group <- ifelse(plot_df$group == 'all_cells', 'consensus', plot_df$group)
plot_df$group <- factor(
    as.character(plot_df$group),
    levels = gsub('-', '', wgcna_groups)
)
plot_df$group <- fct_rev(plot_df$group)

plot_df$Diagnosis <- factor(
    as.character(plot_df$Diagnosis), 
    levels = c('Control', 'earlyAD', 'AD', 'AD_DS')
)


mod_levels <- unlist(lapply(wgcna_groups, function(x){
    GetModules(seurat_human, x) %>% .$module %>% levels
}))
mod_levels <- mod_levels[mod_levels != 'grey']

plot_df$module <- factor(as.character(plot_df$module), levels=mod_levels)


#-----------------------------------------------#
# consensus modules
#-----------------------------------------------#

label_df <- plot_df %>% 
    subset(trait == 'scdrs_score' & wgcna == 'consensus') %>% 
    group_by(Diagnosis) %>% 
    slice_max(order_by=cor, n=10)

plot_range <- plot_df %>% subset(trait == 'scdrs_score' & wgcna == 'consensus') %>% .$cor %>% abs %>% max
plot_range <- c(-plot_range + 0.01, plot_range + 0.01)

p1 <- plot_df %>% subset(trait == 'scdrs_score' & wgcna == 'consensus') %>%
    ggplot(aes(x = module, y = group, fill=cor)) + 
    geom_tile() + 
    geom_text(
        data = label_df,
        aes(x = module, y = group, label=round(cor, 2)), size=1.5
    ) +
    scale_fill_gradient2(high='red', mid='white', low='blue', limits=plot_range) + 
    #coord_equal() +
    RotatedAxis() +
    theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(linewidth=1, color='black', fill=NA),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(hjust=0.5),
        panel.grid = element_line(size=0.25, color='lightgrey'),
    ) 



pdf(paste0(fig_dir, 'MT_cor_heatmap_consensus_human.pdf'), width=9, height=3)
p1 + facet_wrap(~Diagnosis, ncol=4)
dev.off()


pdf(paste0(fig_dir, 'MT_cor_heatmap_consensus_human.pdf'), width=4, height=6)
p1 + facet_wrap(~Diagnosis, ncol=1)
dev.off()


#-----------------------------------------------#
# region modules
#-----------------------------------------------#

label_df <- plot_df %>% 
    subset(trait == 'scdrs_score' & wgcna != 'consensus') %>% 
    group_by(Diagnosis) %>% 
    slice_max(order_by=cor, n=10)

plot_range <- plot_df %>% subset(trait == 'scdrs_score' & wgcna != 'consensus') %>% .$cor %>% abs %>% max
plot_range <- c(-plot_range + 0.01, plot_range + 0.01)

p1 <- plot_df %>% subset(trait == 'scdrs_score' & wgcna != 'consensus') %>%
    ggplot(aes(x = module, y = Diagnosis, fill=cor)) + 
    geom_tile() + 
    geom_text(
        data = label_df,
        aes(x = module, y = Diagnosis, label=round(cor, 2)), size=2
    ) +
    scale_fill_gradient2(high='red', mid='white', low='blue', limits=plot_range) + 
    #coord_equal() 
    RotatedAxis() +
    theme(
        axis.line.x = element_blank(),
        axis.line.y = element_blank(),
        panel.border = element_rect(linewidth=1, color='black', fill=NA),
        axis.title.x = element_blank(),
        axis.title.y = element_blank(),
        plot.title = element_text(hjust=0.5),
        panel.grid = element_line(size=0.25, color='lightgrey'),
    ) 



pdf(paste0(fig_dir, 'MT_cor_heatmap_allmods_human.pdf'), width=8, height=12)
p1 + facet_wrap(~group, scales='free', ncol=1)
dev.off()


```





Make a DotPlot where the dot color is scDRS correlation 
and the size is the average ME 

Consensus modules 

```{r eval=FALSE}

# what group shad a significant scDRS?
signif_groups <- c('L1', 'L34', 'WM')

plot_df <- read.csv(file=paste0(data_dir, 'scDRS_module_correlation.csv'))

# only consensus for now 
plot_df <- subset(plot_df, wgcna == 'consensus' & group != 'all_cells' & trait == 'scdrs_score')

plot_df$group <- factor(
    as.character(plot_df$group),
    levels = c(gsub('-', '', names(human_cp)))
)

table(plot_df$group)

plot_df$Diagnosis <- factor(
    as.character(plot_df$Diagnosis), 
    levels = c('Control', 'earlyAD', 'AD', 'AD_DS')
)

modules <- GetModules(seurat_human, 'consensus')
mod_levels <- levels(modules$module)

plot_df %>% arrange(-cor) %>% head(25)

plot_df$module <- factor(as.character(plot_df$module), levels=mod_levels)

# grouping variables:
plot_df$id <- paste(plot_df$group, plot_df$Diagnosis, sep='_')

# get wgcna info 
data.features <- GetMEs(seurat_human, wgcna_name='consensus') %>% as.data.frame()
data.features$id <- seurat_human$anno_DMEs


#---------------------------------------------------------------------
# code from Seurat DotPlot
#---------------------------------------------------------------------

split.by <- 'Diagnosis'
scale <- TRUE
col.min <- -2.5
col.max <- 2.5
dot.min <- 0

data.features$id <- paste(seurat_human$anno_DMEs, seurat_human$Diagnosis, sep='_')

data.plot <- lapply(
    X = unique(x = data.features$id),
    FUN = function(ident) {
      data.use <- data.features[data.features$id == ident, 1:(ncol(x = data.features) - 1), drop = FALSE]
      avg.exp <- apply(
        X = data.use,
        MARGIN = 2,
        FUN = function(x) {
          return(mean(x = expm1(x = x)))
        }
      )
      pct.exp <- apply(X = data.use, MARGIN = 2, FUN = PercentAbove, threshold = 0)
      return(list(avg.exp = avg.exp, pct.exp = pct.exp))
    }
  )
names(x = data.plot) <- unique(x = data.features$id)
id.levels <- levels(seurat_obj$cell_type)
data.plot <- lapply(
    X = names(x = data.plot),
    FUN = function(x) {
      data.use <- as.data.frame(x = data.plot[[x]])
      data.use$features.plot <- rownames(x = data.use)
      data.use$id <- x
      return(data.use)
    }
  )
  data.plot <- do.call(what = 'rbind', args = data.plot)
  if (!is.null(x = id.levels)) {
    data.plot$id <- factor(x = data.plot$id, levels = id.levels)
  }
  ngroup <- length(x = levels(x = data.plot$id))

  avg.exp.scaled <- sapply(
    X = unique(x = data.plot$features.plot),
    FUN = function(x) {
      data.use <- data.plot[data.plot$features.plot == x, 'avg.exp']
      if (scale) {
        data.use <- scale(x = log1p(data.use))
        data.use <- MinMax(data = data.use, min = col.min, max = col.max)
      } else {
        data.use <- log1p(x = data.use)
      }
      return(data.use)
    }
  )
  avg.exp.scaled <- as.vector(x = t(x = avg.exp.scaled))
  data.plot$avg.exp.scaled <- avg.exp.scaled
  
  data.plot$pct.exp[data.plot$pct.exp < dot.min] <- NA
  data.plot$pct.exp <- data.plot$pct.exp * 100

# match with the other table
data.plot <- subset(data.plot, id %in% plot_df$id)

# merge the DFs
data.plot$ix <- paste0(data.plot$features.plot, '_', data.plot$id)
plot_df$ix <- paste0(plot_df$module, '_', plot_df$id)

ix <- match(plot_df$ix, data.plot$ix, )
all.equal(plot_df$ix, data.plot$ix[ix])

plot_df$avg.exp <- data.plot$avg.exp[ix]
plot_df$avg.exp.scaled <- data.plot$avg.exp.scaled[ix]
plot_df$pct.exp <- data.plot$pct.exp[ix]



plot_range <- plot_df %>% subset(group %in% signif_groups) %>% .$cor %>% range
size_range <- plot_df %>% subset(group %in% signif_groups) %>% .$pct.exp %>% range

# loop through Studies & diagnosis
diagnoses <- unique(as.character(seurat_human$Diagnosis))
plot_list <- list()
for(cur_diag in diagnoses){
 
    cur_df <- plot_df %>% subset(group %in% signif_groups & Diagnosis == cur_diag)
    p <-cur_df %>%
        ggplot(aes(x = module, y = group, color=cor, size=pct.exp)) + 
        geom_point() + 
        scale_color_gradient2(high='red', mid='white', low='blue', limits=plot_range) + 
        scale_size_continuous(limits=size_range) +
        coord_equal() +
        RotatedAxis() +
        ylab(paste0(cur_diag)) + 
        theme(
            axis.line.x = element_blank(),
            axis.line.y = element_blank(),
            panel.border = element_rect(linewidth=1, color='black', fill=NA),
            axis.title.x = element_blank(),
            axis.title.y = element_text(angle=0, vjust=0.5),
            plot.title = element_text(hjust=0.5),
            panel.grid = element_line(size=0.25, color='lightgrey'),
            plot.margin = margin(c(0,0,0,0)),
            
        ) 

    if(length(plot_list) > 0){
        p <- p + theme(
            axis.text.x = element_blank(),
            axis.ticks.x = element_blank(),
        )
    }

    plot_list[[cur_diag]] <- p

}


pdf(paste0(fig_dir, 'MT_cor_heatmap_consensus_visium_dotplot.pdf'), width=6, height=5)
wrap_plots(rev(plot_list), ncol=1) + plot_layout(guides='collect')
dev.off()



```

All of the other modules 

```{r eval=FALSE}

# what group shad a significant scDRS?
signif_groups <- c('L1', 'L34', 'WM')

plot_df <- read.csv(file=paste0(data_dir, 'scDRS_module_correlation.csv'))
plot_df <- subset(plot_df, wgcna != 'consensus' & group != 'all_cells' & trait == 'scdrs_score')

tmp2 <- subset(plot_df, wgcna != 'consensus')
ix <- do.call(rbind, strsplit(tmp2$module, '-'))[,1]
tmp2 <- tmp2[ix == gsub('-', '', tmp2$group),]
plot_df <- tmp2

plot_df$group <- factor(
    as.character(plot_df$group),
    levels = c(gsub('-', '', names(human_cp)))
)

table(plot_df$group)

plot_df$Diagnosis <- factor(
    as.character(plot_df$Diagnosis), 
    levels = c('Control', 'earlyAD', 'AD', 'AD_DS')
)

mod_levels <- unlist(lapply(wgcna_groups, function(x){
    GetModules(seurat_human, x) %>% .$module %>% levels
}))
mod_levels <- mod_levels[mod_levels != 'grey']

plot_df$module <- droplevels(factor(as.character(plot_df$module), levels=mod_levels))

# grouping variables:
plot_df$id <- paste(plot_df$group, plot_df$Diagnosis, sep='_')

# get wgcna info 
data.features <- do.call(cbind, lapply(wgcna_groups[wgcna_groups != 'consensus'], function(x){
    GetMEs(seurat_human, wgcna_name=x) %>% as.data.frame()
})) 
data.features$id <- seurat_human$anno_DMEs


#---------------------------------------------------------------------
# code from Seurat DotPlot
#---------------------------------------------------------------------

split.by <- 'Diagnosis'
scale <- TRUE
col.min <- -2.5
col.max <- 2.5
dot.min <- 0

data.features$id <- paste(seurat_human$anno_DMEs, seurat_human$Diagnosis, sep='_')

data.plot <- lapply(
    X = unique(x = data.features$id),
    FUN = function(ident) {
      data.use <- data.features[data.features$id == ident, 1:(ncol(x = data.features) - 1), drop = FALSE]
      avg.exp <- apply(
        X = data.use,
        MARGIN = 2,
        FUN = function(x) {
          return(mean(x = expm1(x = x)))
        }
      )
      pct.exp <- apply(X = data.use, MARGIN = 2, FUN = PercentAbove, threshold = 0)
      return(list(avg.exp = avg.exp, pct.exp = pct.exp))
    }
  )
names(x = data.plot) <- unique(x = data.features$id)
id.levels <- levels(seurat_obj$cell_type)
data.plot <- lapply(
    X = names(x = data.plot),
    FUN = function(x) {
      data.use <- as.data.frame(x = data.plot[[x]])
      data.use$features.plot <- rownames(x = data.use)
      data.use$id <- x
      return(data.use)
    }
  )
  data.plot <- do.call(what = 'rbind', args = data.plot)
  if (!is.null(x = id.levels)) {
    data.plot$id <- factor(x = data.plot$id, levels = id.levels)
  }
  ngroup <- length(x = levels(x = data.plot$id))

  avg.exp.scaled <- sapply(
    X = unique(x = data.plot$features.plot),
    FUN = function(x) {
      data.use <- data.plot[data.plot$features.plot == x, 'avg.exp']
      if (scale) {
        data.use <- scale(x = log1p(data.use))
        data.use <- MinMax(data = data.use, min = col.min, max = col.max)
      } else {
        data.use <- log1p(x = data.use)
      }
      return(data.use)
    }
  )
  avg.exp.scaled <- as.vector(x = t(x = avg.exp.scaled))
  data.plot$avg.exp.scaled <- avg.exp.scaled
  
  data.plot$pct.exp[data.plot$pct.exp < dot.min] <- NA
  data.plot$pct.exp <- data.plot$pct.exp * 100

# match with the other table
data.plot <- subset(data.plot, id %in% plot_df$id)

# merge the DFs
data.plot$ix <- paste0(data.plot$features.plot, '_', data.plot$id)
plot_df$ix <- paste0(plot_df$module, '_', plot_df$id)

ix <- match(plot_df$ix, data.plot$ix, )
all.equal(plot_df$ix, data.plot$ix[ix])

plot_df$avg.exp <- data.plot$avg.exp[ix]
plot_df$avg.exp.scaled <- data.plot$avg.exp.scaled[ix]
plot_df$pct.exp <- data.plot$pct.exp[ix]



plot_range <- plot_df %>% subset(group %in% signif_groups) %>% .$cor %>% range
size_range <- plot_df %>% subset(group %in% signif_groups) %>% .$pct.exp %>% range

# loop through Studies & diagnosis
plot_list <- list()
for(cur_wgcna in wgcna_groups[wgcna_groups != 'consensus']){
 
    cur_df <- plot_df %>% subset(group %in% signif_groups & wgcna == cur_wgcna)
    
    if(nrow(cur_df) == 0){next}
   
    p <-cur_df %>%
        ggplot(aes(x = module, y = Diagnosis, color=cor, size=pct.exp)) + 
        geom_point() + 
        scale_color_gradient2(high='red', mid='white', low='blue', limits=plot_range) + 
        scale_size_continuous(limits=size_range) +
        coord_equal() +
        RotatedAxis() +
        ylab(paste0(cur_wgcna)) + 
        theme(
            axis.line.x = element_blank(),
            axis.line.y = element_blank(),
            panel.border = element_rect(linewidth=1, color='black', fill=NA),
            axis.title.x = element_blank(),
            axis.title.y = element_text(angle=0, vjust=0.5),
            plot.title = element_text(hjust=0.5),
            panel.grid = element_line(size=0.25, color='lightgrey')
        ) 


    plot_list[[cur_wgcna]] <- p

}


pdf(paste0(fig_dir, 'MT_cor_heatmap_consensus_visium_allmods_dotplot.pdf'), width=8, height=6)
wrap_plots(rev(plot_list), ncol=1) + plot_layout(guides='collect')
dev.off()



```




NOT RUN BELOW:


Try to get the Moran's I or the Geary's C to work from the Voyager package:

Following this tutorial from Voyager:
https://pachterlab.github.io/voyager/articles/nonspatial.html

```{r eval=FALSE}


library(Voyager)
library(SpatialFeatureExperiment)
library(SpatialExperiment)
library(DropletUtils)
library(BiocNeighbors)
library(scater)
library(scran)
library(bluster)
library(BiocParallel)
library(scuttle)
library(stringr)
library(BiocSingular)
library(spdep)
library(patchwork)
library(dplyr)
library(reticulate)

# seurat_morabito <- readRDS("/dfs7/swaruplab/smorabit/analysis/scWGCNA/data/Swarup_2021.rds")

# just get the Morabito / Miyoshi 2021 dataset 
cur_seurat <- subset(seurat_human, Study == 'ADDS')

# load the scDRS results
scdrs_df <- read.csv(file=paste0(data_dir,'ADDS_scDRS_results.csv'))

# subset for one trait
cur_df <- subset(scdrs_df, trait == 'Alzheimers_Jansen2019')

# make sure to only use barcodes that match between the Seurat obj and the scDRS 
cur_seurat <- subset(seurat_human, bc %in% cur_df$barcode)

ix <- match(cur_seurat$bc, cur_df$barcode)
all.equal(as.character(cur_df$barcode)[ix], as.character(cur_seurat$bc))

cur_seurat$scdrs_score <- cur_df[ix,]$norm_score
cur_seurat$scdrs_zscore <- cur_df[ix,]$zscore
cur_seurat$scdrs_pval <- cur_df[ix,]$pval


# harmony <- seurat_morabito@reductions$harmony@cell.embeddings
# harmony <- harmony[cur_seurat$bc,]
# rownames(harmony) <- colnames(cur_seurat)

# cur_seurat@reductions$harmony <- Seurat::CreateDimReducObject(
#   embeddings = harmony,
#   key="harmony",
#   assay="RNA"
# )




p1 <- VlnPlot(
    cur_seurat, 
    features = 'scdrs_score',
    group.by = 'cell_identity', 
    pt.size=0) + NoLegend()

p2 <- VlnPlot(
    cur_seurat, 
    features = 'scdrs_zscore',
    group.by = 'cell_identity', 
    pt.size=0) + NoLegend()


pdf(paste0(fig_dir, 'test_scdrs_vln_ADDS_AD.pdf'), width=10, height=6)
p1 / p2
dev.off()


#cur_seurat <- subset(seurat_human, Study == 'Morabito_Miyoshi')

# convert it to SingleCellExperiment
sce <- as.SingleCellExperiment(cur_seurat)

#-------------------------------------------------------------#
# Data processing
#-------------------------------------------------------------#

sce <- logNormCounts(sce)

#-------------------------------------------------------------#
# Run KNN on the SCVANI representation
#-------------------------------------------------------------#

# run KNN
foo <- findKNN(reducedDim(sce, "SCANVI")[,1:30], k=10, BNPARAM=AnnoyParam())
#foo <- findKNN(reducedDim(sce, "HARMONY")[,1:30], k=10, BNPARAM=AnnoyParam())

# Split by row
foo_nb <- asplit(foo$index, 1)
dmat <- 1/foo$distance

# Row normalize the weights
dmat <- sweep(dmat, 1, rowSums(dmat), FUN = "/")
glist <- asplit(dmat, 1)

# Sort based on index
ord <- lapply(foo_nb, order)
foo_nb <- lapply(seq_along(foo_nb), function(i) foo_nb[[i]][ord[[i]]])
class(foo_nb) <- "nb"
glist <- lapply(seq_along(glist), function(i) glist[[i]][ord[[i]]])

listw <- list(style = "W",
              neighbours = foo_nb,
              weights = glist)
class(listw) <- "listw"
attr(listw, "region.id") <- colnames(sce)

#-------------------------------------------------------------#
# Convert formats from SCE to SFE
#-------------------------------------------------------------#

# convert to SFE format
(sfe <- toSpatialFeatureExperiment(sce, spatialCoords = reducedDim(sce, "UMAP")[,1:2],
                                  spatialCoordsNames = NULL))

# add the knn to the SFE
colGraph(sfe, "knn") <- listw

#-------------------------------------------------------------#
# Test running moran's I
#-------------------------------------------------------------#

# why nans?
sfe <- runMoransI(sfe,  
    features = c('CSF1R', 'SLC17A7', 'MOBP'))
rowData(sfe)[c('CSF1R', 'SLC17A7', 'MOBP'),]


# gloabl Moran's I
sfe <- colDataMoransI(sfe, c("nCount_RNA", "doublet_scores", 'scdrs_score'))
colFeatureData(sfe)[ c("nCount_RNA", "doublet_scores", 'scdrs_score'),]

sfe <- colDataUnivariate(sfe, "moran.plot",  c("nCount_RNA", "doublet_scores", 'scdrs_score'))

p <- moranPlot(sfe, "scdrs_score", color_by = "cell_identity")

pdf(paste0(fig_dir, 'test_voyager_moranI_scdrs_score.pdf'), width=10, height=6)
p
dev.off()

# Local Moran's I
sfe <- colDataUnivariate(sfe, "localmoran",  c("nCount_RNA", "doublet_scores"))

pdf(paste0(fig_dir, 'test_voyager_local_moranI.pdf'), width=10, height=6)
plotSpatialFeature(sfe,c("nCount_RNA", "doublet_scores"))
dev.off()

pdf(paste0(fig_dir, 'test_voyager_local_moranI.pdf'), width=10, height=6)
plotLocalResult(sfe, "localmoran",  c("nCount_RNA", "doublet_scores"), 
                colGeometryName = "centroids",
                divergent = TRUE, diverge_center = 0, ncol = 2)
dev.off()

#-------------------------------------------------------------#
# Try it with Geary's C (not in the tutorial)
#-------------------------------------------------------------#

# Geary's C
sfe <- colDataUnivariate(sfe, type="geary",  c("nCount_RNA", "doublet_scores"))

# LOSH
sfe <- colDataUnivariate(sfe, type="LOSH",  c("nCount_RNA", "doublet_scores"))


colFeatureData(sfe)[c("nCount_RNA", "doublet_scores"),]


#-------------------------------------------------------------#
# Can I run Moran's I on just one cell cluster?
# 
# Yes you can. But you first have to make a new KNN graph after 
# subsetting.
#-------------------------------------------------------------#

# sfe <- colDataUnivariate(sfe, "moran",  c("nCount_RNA", "doublet_scores"), sample_id = 'cell_type')
# colFeatureData(sfe)[ c("nCount_RNA", "doublet_scores"),]

test <- sfe[,sfe$cell_type == 'MG']

# run KNN
foo <- findKNN(reducedDim(test, "SCANVI")[,1:30], k=25, BNPARAM=AnnoyParam())

# Split by row
foo_nb <- asplit(foo$index, 1)
dmat <- 1/foo$distance

# Row normalize the weights
dmat <- sweep(dmat, 1, rowSums(dmat), FUN = "/")
glist <- asplit(dmat, 1)

# Sort based on index
ord <- lapply(foo_nb, order)
foo_nb <- lapply(seq_along(foo_nb), function(i) foo_nb[[i]][ord[[i]]])
class(foo_nb) <- "nb"
glist <- lapply(seq_along(glist), function(i) glist[[i]][ord[[i]]])

listw <- list(style = "W",
              neighbours = foo_nb,
              weights = glist)
class(listw) <- "listw"
attr(listw, "region.id") <- colnames(test)

colGraph(test, "knn") <- listw


test <- colDataUnivariate(test, "moran",  c("nCount_RNA", "doublet_scores", 'scdrs_score'))
colFeatureData(test)[ c("nCount_RNA", "doublet_scores", 'scdrs_score'),]

test <- colDataUnivariate(test, "geary",  c("nCount_RNA", "doublet_scores", 'scdrs_score'))
colFeatureData(test)[ c("nCount_RNA", "doublet_scores", 'scdrs_score'),]

test <- runMoransI(test,  
    features = c('CSF1R', 'SLC17A7', 'MOBP'))
rowData(test)[c('CSF1R', 'SLC17A7', 'MOBP'),]

# run this for all genes:
test <- runMoransI(test, features = rownames(test), BPPARAM = MulticoreParam(2))

p <- plotRowDataHistogram(test, "moran_sample01", bins = 50) +
    scale_color_continuous(breaks = scales::breaks_width(2))


pdf(paste0(fig_dir, 'test_voyager_moranI_hist.pdf'), width=10, height=6)
p
dev.off()

```